---
title: Analysis of Tfh response in context of mRNA vaccination.
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "March 17, 2023"
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Using scRepetoire}
  %\usepackage[UTF-8]{inputenc}
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Loading Libraries

```{r}
suppressPackageStartupMessages({
      library(bluster)
      library(dplyr)
      library(eulerr)
      library(escape)
      library(ggalluvial)
      library(ggseqlogo)
      library(ggplot2)
      library(ggrepel)
      library(harmony)
      library(iNEXT)
      library(leidenAlg)
      library(Matrix)
      library(msa)
      library(mumosa)
      library(Nebulosa)
      library(patchwork)
      library(phateR)
      library(ProjecTILs)
      library(RColorBrewer)
      library(scRepertoire)
      library(seqinr)
      library(Seurat)
      library(SingleR)
      library(stringdist)
      library(stringr)
      library(Trex)
})
```

# Processing Data

General pipeline for the generating intermediate Seurat objects that have been 1) filtered/passe qc and 2) annotated with multiple packages

```{r eval=FALSE}
dir.create("dataExplore")
dir.create("dataExplore/qc")
"%!in%" <- Negate("%in%")
file_list <- list.files("./data/")



#############################
#Loading Annotation Data Sets
############################

DICE <- DatabaseImmuneCellExpressionData()
ref8 <- load.reference.map("./data/annotations/ref_LCMV_Atlas_mouse_v1.rds")
ref4 <- load.reference.map("./data/annotations/ref_LCMV_CD4_mouse_release_v1.rds")
    
for (i in seq_along(file_list)){
    tmp <-  Read10X(paste0("./data/sequencingRuns/", file_list[i], "/filtered_feature_bc_matrix"))
    s.obj <- CreateSeuratObject(tmp, project = file_list[i])
    rm(tmp)
    s.obj <- RenameCells(object = s.obj, new.names = paste0(file_list[i], "_", rownames(s.obj[[]])))
    s.obj<- subset(s.obj, subset = nFeature_RNA > 100)
    s.obj[["mito.genes"]] <- PercentageFeatureSet(s.obj, pattern = "^MT-")
    
    p1 <- VlnPlot(object = s.obj, features = c("nCount_RNA")) + theme(legend.position = "none")
    p2 <- VlnPlot(object = s.obj, features = c("nFeature_RNA")) + theme(legend.position = "none")
    p3 <- VlnPlot(object = s.obj, features = c("mito.genes")) + theme(legend.position = "none")
    
    pdf(paste0("./dataExplore/qc/", file_list[i], ".pdf"), height = 8, width=12)
    grid.arrange(p1, p2, p3, ncol = 3)
    dev.off()
    
  ###########################
  #Here is the filtering step
  ############################
  standev <- sd(log(s.obj$nFeature_RNA))*2 #cutting off above standard deviation of 2
  mean <- mean(log(s.obj$nFeature_RNA))
  cut <- round(exp(standev+mean))
  
  p4 <- FeatureScatter(s.obj, feature1 = "nCount_RNA", feature2 = "mito.genes") + 
      geom_hline(yintercept = 15)
  p5 <- FeatureScatter(s.obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + 
      geom_hline(yintercept = cut)
   
  pdf(paste0("./dataExplore/qc/", file_list[i], "_scatterPlot.pdf"), height = 4, width=8)
    grid.arrange(p4, p5, ncol = 2)
  dev.off()
  
  s.obj <- subset(s.obj, subset = mito.genes < 15 & nFeature_RNA < cut)
    
  ###########################################
  #Estimate Doublets for Each Sequencing Run
  ############################################
  sce <- as.SingleCellExperiment(s.obj)
  sce <- scDblFinder(sce, BPPARAM=MulticoreParam(3))
  doublets <- data.frame(db.weight.score = sce$scDblFinder.score, db.ratio = sce$scDblFinder.weighted, 
                         db.class = sce$scDblFinder.class, db.score = sce$scDblFinder.score)
  rownames(doublets) <- rownames(sce@colData)
  s.obj <- AddMetaData(s.obj, doublets)
  rm(sce)
  
  ## Finding variable features and normalizing data for integration to follow
  s.obj <- NormalizeData(s.obj)
  s.obj <- FindVariableFeatures(s.obj, selection.method = "vst", nfeatures = 2000)
  s.obj <- quietTCRgenes(s.obj)
  
  #############################################
  #ProjecTIL Annotation of Cell Types
  #############################################
  
    query.projected <- make.projection(s.obj, ref = ref4, ncores = 1)
    
    #Lapply across list of Seurat objects
    query.projected <- cellstate.predict(ref = ref4, query =query.projected , reduction = "umap", ndim = 2)
      meta <- query.projected[[]][c("functional.cluster", "functional.cluster.conf")]
      colnames(meta) <- c("CD4.annot", "CD4.score")
      rownames(meta) <- stringr::str_remove(rownames(meta), "Q_")
    s.obj <- AddMetaData(s.obj, meta)
    rm(query.projected)

    ###########################
    #CD8 LCMV Annotation
    ############################
    
    query.projected <- make.projection(s.obj, ref = ref8, ncores = 1)
    
    query.projected <- cellstate.predict(ref = ref8, query = query.projected, reduction = "umap", ndim = 2)
    meta <- query.projected[[]][c("functional.cluster", "functional.cluster.conf")]
    colnames(meta) <- c("CD8.annot", "CD8.score")
    rownames(meta) <- stringr::str_remove(rownames(meta), "Q_")
    s.obj <- AddMetaData(s.obj, meta)
    rm(query.projected)
    rm(ref)
    
    
    DICE.annotation <- NULL
    tmp.2 <- s.obj@assays[["RNA"]]@counts
    ####This approach for matrix conversion saves some memory
    tmp.2 <- tmp.2[tabulate(summary(tmp.2)$i) != 0, , drop = FALSE]
    com.res2 <- SingleR(tmp.2, ref=DICE, labels=DICE$label.fine, assay.type.test=1)
      df <- data.frame("DICE.labels" = com.res2$labels, "DICE.pruned.labels" = com.res2$pruned.labels)
      rownames(df) <- rownames(s.obj[[]])
      s.obj <- AddMetaData(s.obj, df)
 saveRDS(s.obj, file = paste0("./data/", file_list[i], "/processedRNA_SeuratObject.rds"))
 rm(s.obj)
}
```


## Integrating Data


```{r, eval = FALSE}
inventory <- read.delim("./data/data.inventory.txt")
samples<- inventory$donor

list <- list()
for (i in seq_along(samples)) {
  list[[i]] <- readRDS(paste0("./data/sequencingRuns/", samples[i], "/processedRNA_SeuratObject.rds"))
}

#Merging Data
list <- merge(list[[1]], list[-1])

data.inventory <- read.delim("./data/data.inventory.txt")
meta <- list[[]]
#meta <- meta[,-which(colnames(meta) %in% colnames(data.inventory))]
barcodes <- rownames(meta)
meta <- merge(meta, data.inventory, by.x = "orig.ident", by.y = "sample")
meta <- meta[,colnames(data.inventory)[-2]]
rownames(meta) <- barcodes

list <- AddMetaData(list, meta)


list <- list %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(verbose = FALSE) %>% 
    RunPCA(pc.genes = pbmc@var.genes, npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident")) %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:25,
            n.epochs = 500) %>%
    FindClusters(resolution = 1, 
                 algorithm = 3) %>% 
    identity()

saveRDS(list, file = "./data/processedData/Fullintegrated_seuratObjects.rds")
```


## Adding Contig Data


```{r, eval = FALSE}
######################################
#iterate to make a list of contig csvs
######################################

contig.list <- NULL
for (i in seq_along(samples)){
  contig.list[[i]] <-  read.csv(paste0("./data/sequencingRuns", samples[i], "/filtered_contig_annotations.csv"))
}
names(contig.list) <- samples

##################################################
#Reducing the data to the individual barcode level
##################################################
combinedObject <- combineTCR(contig.list, samples = samples, filterMulti = TRUE, removeNA = TRUE, cells = "T-AB")

pt <- data.inventory$donor[match(samples, data.inventory$sample)]
combinedObject <- addVariable(combinedObject, name = "Patient", variables = pt)
list <- combineExpression(combinedObject, list, cloneCall = "aa", group.by = "Patient")

slot(list, "meta.data")$cloneType <- factor(slot(list, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (0.1 < X <= 1)", 
                           "Large (0.01 < X <= 0.1)", 
                            "Medium (0.001 < X <= 0.01)", 
                            "Small (1e-04 < X <= 0.001)", 
                            "Rare (0 < X <= 1e-04)", NA))

DimPlot(list, group.by = "cloneType") + scale_color_manual(values = rev(viridis_pal(option = "B")(length(unique(list$cloneType)))))

saveRDS(list, file = "./data/processedData/Fullintegrated_seuratObjects.rds")
```

## Isolating Only T Cells

```{r, eval=FALSE}
#Filtering using the scGate built into ProjecTIL based on T-AB
#Only T cells with contig information
list$filter  <- "Yes"
list$filter <- ifelse(!is.na(list$CD8.annot) | !is.na(list$CD4.annot), "No", list$filter)
list$filter <- ifelse(is.na(list$CTaa), "Yes", list$filter)

list <- subset(list, filter != "Yes")

list@meta.data$CD8.pos <- ifelse(list@assays$RNA@data["CD8A",] >=  0.4, "Yes", "No")

dim(list)[2]
#Check: 219283 cells

list <- list %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(verbose = FALSE) %>% 
    RunPCA(pc.genes = pbmc@var.genes, npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident", "tissue"), 
               max.iter.harmony = 15) %>%
    FindNeighbors(reduction = "harmony", dims = 1:15) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:15,
            n.epochs = 500) %>%
    FindClusters(resolution = 0.8, 
                 algorithm = 3) %>% 
    identity()

saveRDS(list, file = "./data/processedData/Fullintegrated_seuratObjects_Tcells.rds")


#Relevel the Clonotype Expansion assignments
meta <- list[[]]
meta <- split(meta, meta$donor)

cloneTypes=c(None = 0, Rare = 1e-5, Small = 1e-4, 
                              Medium = 1e-3, Large = 1e-2, Hyperexpanded = 1)

Con.df <- NULL
for(i in seq_along(meta)) {
  tmp <- meta[[i]]
  CDstatus <- table(tmp$CD8.pos)
  tmp <- tmp %>% group_by(tmp[,"CTaa"], CD8.pos) %>%
                      summarise(Clone.Prop = n())
  tmp$Clone.Prop[which(tmp$CD8.pos == "No")] <- tmp$Clone.Prop[tmp$CD8.pos == "No"]/CDstatus[1]
  tmp$Clone.Prop[which(tmp$CD8.pos == "Yes")] <- tmp$Clone.Prop[tmp$CD8.pos == "Yes"]/CDstatus[2]
  colnames(tmp)[1] <- "CTaa"
  data.CD4 <- merge(meta[[i]][meta[[i]]$CD8.pos == "No",], tmp[which(tmp$CD8.pos == "No"),], by = "CTaa")
  data.CD8 <- merge(meta[[i]][meta[[i]]$CD8.pos == "Yes",], tmp[which(tmp$CD8.pos == "Yes"),], by = "CTaa")
  data <- rbind.data.frame(data.CD4, data.CD8)
  data <- data[,c("barcode", "Clone.Prop")]
  Con.df <- rbind.data.frame(Con.df, data)
}

names.cell <- Con.df$barcode
Con.df <- as.data.frame(Con.df[,-1])
rownames(Con.df) <- names.cell
colnames(Con.df) <- "Clone.Prop"
Con.df$Clone.Group <- NA

for (x in seq_along(cloneTypes)) { names(cloneTypes)[x] <- 
        paste0(names(cloneTypes[x]), ' (', cloneTypes[x-1], 
        ' < X <= ', cloneTypes[x], ')') }
    for (i in 2:length(cloneTypes)) { 
      Con.df$Clone.Group <- ifelse(Con.df$Clone.Prop > cloneTypes[i-1] & 
                                Con.df$Clone.Prop <= cloneTypes[i], 
                                names(cloneTypes[i]), Con.df$Clone.Group) 
      }

list <- AddMetaData(list, Con.df)
```

## Adding Known Epitope Sequences Annotation

```{r, eval = FALSE}
#Spike 167-180 epitope
sequence <- read.csv("./sequenceInfo/Sequences.csv")
CTaa <- as.data.frame(stringr::str_split(list$CTaa, "_", simplify = TRUE))
TA <-which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

list$TCRB.S167.180<- NA
list$TCRB.S167.180[TB] <- "Specific"

list$TCRA.S167.180 <- NA
list$TCRA.S167.180[TA] <- "Specific"


#Spike 815-830 epitope

sequence <- read.csv("./sequenceInfo/dykema.sequences.csv")
CTaa <- as.data.frame(stringr::str_split(list$CTaa, "_", simplify = TRUE))
TA <-which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

list$TCRB.S815.830 <- NA
list$TCRB.S815.830[TB] <- "Specific"

list$TCRA.S815.830 <- NA
list$TCRA.S815.830[TA] <- "Specific"

# VDJdb sequence

sequences <- read.delim("./sequenceInfo/VDJdb.tsv")
sequences <- sequences[sequences$MHC.class == "MHCII" & sequences$Epitope.gene == "Spike",]

epitope <- c("NCTFEYVSQPFLMDL" = "S164.179",
"RISNCVADYSVLYNS" = "S356.371",
"NFSQILPDPSKPSKR" = "S800.815",
"HWFVTQRNFYEPQII" = "S1101.1115",
"VGGNYNYLYRLFRKS" = "S445.459",
"NNSYECDIPIGAGIC" = "S657.671")

list$TCRB.VDJ.db <- NA
list$TCRA.VDJ.db <- NA
for (i in seq_along(epitope)) {
  sequences$Epitope[which(sequences$Epitope == names(epitope)[i])] <- epitope[i]
  TA <- sequences[sequences$Gene == "TRA" & sequences$Epitope == epitope[i], ]
  TB <- sequences[sequences$Gene == "TRB" & sequences$Epitope == epitope[i], ]
  TA.loc <-which(CTaa$V1 %in% TA$CDR3)
  TB.loc <-which(CTaa$V2 %in% TB$CDR3)
  list$TCRB.VDJ.db[TB.loc] <- epitope[i]
  list$TCRA.VDJ.db[TA.loc] <- epitope[i]
}


list$TCRB.epitope  <- NA
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.S167.180), "S167.180", NA)
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.S815.830), "S815.830", list$TCRB.epitope)
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.VDJ.db) & is.na(list$TCRB.epitope), list$TCRB.VDJ.db, list$TCRB.epitope)

list$TCRA.epitope  <- NA
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.S167.180), "S167.180", NA)
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.S815.830), "S815.830", list$TCRA.epitope)
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.VDJ.db) & is.na(list$TCRA.epitope), list$TCRA.VDJ.db, list$TCRA.epitope)

saveRDS(list, file = "./processedData/Fullintegrated_seuratObjects_Tcells.rds")
```

# Figure 1 Graphs

```{r}
dir.create("./output/Figure1/")
list <- readRDS("./data/processedData/Fullintegrated_seuratObjects_Tcells.rds")
```


## Figure 1C
```{r}
mycolors <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(list$seurat_clusters)))
DimPlot(list, label = TRUE, raster=FALSE, pt.size = 0.5, repel = TRUE) + 
  theme_void() + 
  scale_color_manual(values = mycolors) 

ggsave("./output/Figure1/Figure1C.pdf", height = 4, width = 4.5)
```

## Figure 1D

Examining the general markers of T cell subtypes: 

```{r}
source("./R/utils.R")


genes <- c("CD4", "CD8A", "CCR7", "SELL", "CXCR5", "ICOS", "FOXP3", "IL2RA", "CTLA4", "PDCD1")

x <- plot_density(list,
                  features = genes, 
                  size = 0.5,
                  combine = FALSE)

for(i in seq_along(x)) {
  x[[i]] <- x[[i]] + 
              guides(color = "none") + 
              scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
              theme_void() + 
              theme(plot.title = element_blank(), 
                    plot.margin = unit(c(0.6, 0.15, 0.6, 0.15), "cm"))
}

plot_a_list(x,5,2)
ggsave("./output/Figure1/Figure1D.pdf", height = 5, width = 2)
```
# Figure 1E

Looking at the relative density of cells by timepoint. In order to ensure a unbiased comparison, we randomly downsample cells to compare equal number of cells. 

```{r}
df <- data.frame(list@reductions$umap@cell.embeddings, list@meta.data)

#Sampling to allow for equal comparisons for density
sampled <- df %>%
  group_by(timepoint) %>%
  sample_n(25000)

sampled$timepoint <- factor(sampled$timepoint, levels = c("d28", "d60", "d110", "d201"))

ggplot() + 
  geom_point(data = sampled, aes(x = UMAP_1, y = UMAP_2), stroke = 0, color = "grey") + 
  stat_density_2d(data = sampled, aes(x = UMAP_1, y = UMAP_2, fill = stat(level)), 
                  geom = "polygon", bins = 25) + 
  facet_grid(.~timepoint) + 
  xlim(-8,12) + 
  ylim(-11,7) + 
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
  guides(fill = "none") + 
  theme_void() 

ggsave("./output/Figure1/Figure1E.png", height = 6, width = 28, dpi = 600, bg = "transparent")
```

## Figure 1F

Looking at the relative density of cells by tissue. In order to ensure a unbiased comparison, we randomly downsample cells to compare equal number of cells. 

```{r}
sampled <- df %>%
  group_by(tissue) %>%
  sample_n(90000)

sampled$tissue <- factor(sampled$tissue, levels = c("PB", "LN"))
ggplot() +
  geom_point(data = sampled, aes(x = UMAP_1, y = UMAP_2), stroke = 0, color = "grey") + 
  stat_density_2d(data = subset(sampled, tissue == "LN"), aes(x = UMAP_1, y = UMAP_2, fill = stat(level)), 
                  geom = "polygon", bins = 25) + 
  stat_density_2d(data = subset(sampled, tissue == "PB"), aes(x = UMAP_1, y = UMAP_2, fill = stat(level)), 
                  geom = "polygon", bins = 25) + 
  facet_grid(.~tissue) + 
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
  xlim(-8,12) + 
  ylim(-11,7) + 
  guides(fill = "none") + 
  theme_void()  + 
  theme(plot.title = element_blank(),
        strip.text.x = element_blank())
ggsave("./output/Figure1/Figure1F.png", height = 6, width = 14, dpi = 600, bg = "transparent")
```

## Figure 1G

Looking at the distribution of spike-specific and other COVID19-specific TCRs based on prior experiments. 

```{r}
#Sorting epitopes into either 167-180 or other
other <- c("S356.371" ,"S164.179","S815.830", "S800.815", "S356.371", "S657.671","S1101.1115")

df$epitope <- NA

other <- c("S356.371" ,"S164.179","S815.830", "S800.815", "S356.371", "S657.671","S1101.1115")
df$epitope[c(which(df$TCRB.epitope == "S167.180"), which(df$TCRA.epitope == "S167.180"))] <- "epitope.S167.180"
df$epitope[c(which(df$TCRB.epitope %in% other), which(df$TCRA.epitope %in% other))] <- "other.spike"

ggplot() +
  geom_density2d(data = df, aes(x = UMAP_1, y = UMAP_2), color = "black", lwd = 0.5) + 
  geom_point(data = subset(df, !is.na(epitope)), aes(x = UMAP_1, y = UMAP_2, fill = epitope), shape=21, stroke = 0.25, size = 3) + 
  guides(fill = "none") + 
  theme_void() + 
  scale_fill_manual(values = rev(brewer.pal(11, "RdYlBu")[c(2,10)]))
ggsave("./output/Figure1/Figure1G.png", height = 6, width = 7, dpi = 600, bg = "transparent")
```

## Figure 1H

Generating motifs for the antigen-specific TCRs.

```{r}
meta <- list[[]]
meta <- meta[!is.na(meta$TCRB.S167.180) |!is.na(meta$TCRA.S167.180),] 
TRB <- unique(str_split(meta$CTaa, "_", simplify = TRUE)[,2])
TRA <- unique(str_split(meta$CTaa, "_", simplify = TRUE)[,1])

#################
#S167-180 Epitope
#################

align <- msa(TRA, 
              type = "protein", 
              maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave("./output/Figure1/Figure1H.Specific167.TRA.pdf", height =3, width = 4.5)
  
    align <- msa(TRB, 
               type = "protein", 
               maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave("./output/Figure1/Figure1H.Specific167.TRB.pdf", height =3, width = 4.5)

#####################
#Other Spike Epitopes
#####################
  
meta <- list[[]]
meta <- meta[!is.na(meta$TCRA.epitope) & is.na(meta$TCRA.S167.180) |!is.na(meta$TCRB.epitope) & is.na(meta$TCRB.S167.180),] 
  TRB <- unique(str_split(meta$CTaa, "_", simplify = TRUE)[,2])
  TRA <- unique(str_split(meta$CTaa, "_", simplify = TRUE)[,1])
  
  align <- msa(TRA, 
               type = "protein", 
               maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave("./output/Figure1/Figure1H.Other.TRA.pdf", height =3, width = 4.5)
  
    align <- msa(TRB, 
               type = "protein", 
               maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave(paste0("./output/Figure1/Figure1H.Other.TRB.pdf"), height =3, width = 4.5)
```


# Figure 2 Graphs

Dimensional reduction on the T follicular helper (Tfh) cells. 

```{r}
dir.create("./output/Figure2/")

list <- readRDS("./data/processedData/Fullintegrated_seuratObjects_Tcells.rds")

tfh <- subset(list, tissue == "LN" & CD4.annot %in% c("Tfh_Memory", "Tfh_Effector"))

DefaultAssay(tfh) <- "RNA"

tfh <- tfh %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    quietTCRgenes() %>%
    ScaleData(verbose = FALSE) %>% 
    RunPCA(pc.genes = pbmc@var.genes, npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident")) %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:20,
            n.epochs = 500) %>%
    FindClusters(resolution = 0.5, 
                 algorithm = 3) %>% 
    identity()

saveRDS(tfh, "./data/processedData/Fullintegrated_seuratObjects_TFH_TFHonly.rds")
```

## Figure 2A

UMAP of Tfh cells in lymph node and peripheral blood.

```{r}
tfh <- readRDS("./data/processedData/Fullintegrated_seuratObjects_TFH_TFHonly.rds")
mycolors <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(tfh$seurat_clusters)))

update_geom_defaults("point", list(stroke = 0.25))
DimPlot(tfh, reduction = "umap", label = TRUE, repel = TRUE, label.size = 4, raster=FALSE, pt.size = 0.25) +
  scale_color_manual(values = mycolors) + 
  guides(colors = "none", fill = "none") + 
  theme_void() + 
  NoLegend() 

ggsave("./output/Figure2/Figure2A_TFHonly.pdf", height = 3.5, width = 4, dpi = 600, bg = "transparent")
```

## Figure 2B

General markers of Tfh and other T cell effector markers. 

```{r}
plot <- NULL
features <- sort(c("BCL6", "CXCR5", "CD274", "ICOS", "CCR7","CD40LG","IL21","CXCL13", "IL4", "FOXP3", "IL10", "SELL"))
for (i in seq_along(features)) {

  if(i == 1) {
    plot <- plot_density(tfh,
                features = features[i], 
                size = 0.5) + 
     scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
   guides(fill = "none", color = "none") + 
      theme_void() + 
      theme(plot.title = element_blank(), 
            plot.margin = margin(10,1,1,1))
  } else {
    plot <- plot + plot_density(tfh, 
                features = features[i], 
                size = 0.5) + 
     scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
    guides(fill = "none", color = "none") + 
      theme_void() + 
      theme(plot.title = element_blank(), 
            plot.margin = margin(10,1,1,1))
  }
}

plot + plot_layout(nrow = 3)
ggsave("./output/Figure2/Figure2B_TFHonly.png", height = 5.3, width = 8, dpi = 600, bg = "transparent")
```

## Figure 2C

Finding and displaying the general markers for Tfh clusters.

```{r}
markers <- FindAllMarkers(tfh)
write.csv(markers, "./output/Figure2/Figure2C_tfh_onlymarkers.csv")

top <- markers %>%
  group_by(cluster) %>%
  slice_max(order_by = avg_log2FC, n = 8)

DotPlot(tfh, features = unique(top$gene)) + 
  scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
  scale_y_discrete(limits=rev)+ 
  coord_flip()
ggsave("./output/Figure2/Figure2C_TFHonly.pdf", height = 20, width = 6)
```

## Figure 2D

Identifying and visualizing gene set enirchments for Tfh cells/clusters.

```{r}
ES <- readRDS("./data/ProcessedData/enrichmentResults.rds")

meta <- tfh[[]]
meta$enrichment.groupings <- NA
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(0,1,2,6,7,10))] <- "Memory.TFH"
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(3))] <- "Effector.TFH"
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(4))] <- "TFR"
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(8))] <- "IL10.TFH"
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(11))] <- "Proliferating.TFH"
meta$enrichment.groupings[which(meta$seurat_clusters %in% c(9))] <- "Cytolytic.TFH"
meta <- meta[which(!is.na(meta[,"enrichment.groupings"])),]

ES2 <- ES[rownames(ES) %in% rownames(meta),]
variances <- apply(ES2, 2, var)

pathways <- colnames(ES2)

ES2 <- merge(ES2, meta, by = 0)

sig.results <- getSignificance(ES2, 
                               group = "seurat_clusters", 
                               gene.sets = pathways, 
                               fit = "KW")

sig.results2 <- getSignificance(ES2, 
                               group = "seurat_clusters", 
                               gene.sets = pathways, 
                               fit = "ANOVA")

KW.sig.pathways <- sig.results %>%
                        slice_max(order_by = Chi.square, n = 200) %>%
                        rownames()

ANOVA.sig.pathways <- sig.results2 %>%
                        slice_max(order_by = f.value, n = 200) %>%
                        rownames()

pathways.intersect <- intersect(KW.sig.pathways, ANOVA.sig.pathways)
pathways.intersect <- pathways.intersect[-c(3,4,8,41,47,57,62,73,78,80,85,86,90,99,101,120,125,139,141)]
pathways.intersect <- pathways.intersect[!grepl("REACTOME", pathways.intersect) ]
write(pathways.intersect, "output/Figure2/Figure2_intersected.pathways.txt")

summary <- ES2 %>%
  group_by(seurat_clusters) %>%
  summarise(across(pathways.intersect, median))

normalize <- function(x) {
      (x- min(x)) /(max(x)-min(x))
}

for(i in 2:43) {
  summary[,i] <- normalize(summary[,i])
}
summary <- as.data.frame(summary)

rownames(summary) <- summary$seurat_clusters
colnames(summary) <- stringr::str_remove_all(colnames(summary), "_UCell")

pdf("./output/Figure2/Figure2D.pdf", height = 8, width = 10)
pheatmap::pheatmap(t(summary[,2:43]))
dev.off()
```
## Figure 2E

```{r}
summary <- ES2 %>%
  group_by(seurat_clusters, timepoint) %>%
  summarise(across(pathways.intersect, median))

for(i in 3:44) {
  summary[,i] <- normalize(summary[,i])
}
summary <- as.data.frame(summary)

colnames(summary) <- stringr::str_remove_all(colnames(summary), "_UCell")


melt.summary <- reshape2::melt(summary, id.vars = c("seurat_clusters", "timepoint"))

melt.summary$timepoint <- factor(melt.summary$timepoint, levels = c("d28", "d60", "d110", "d201"))

means <- melt.summary %>%
  group_by(variable) %>%
  summarise(mean = mean(value))

means <- means[order(means$mean),]
pathways <- rev(means$variable)


melt.summary$order <- NA
for(i in seq_along(pathways)) {
  melt.summary$order[which(melt.summary$variable == pathways[i])] <- i
}


ggplot(melt.summary, aes(x=timepoint, y = value)) + 
  geom_line(aes(color = order, group = interaction(variable, seurat_clusters)), lwd = 0.25) + 
  geom_point(aes(group = interaction(variable, seurat_clusters)), size = 0.5)+ 
  stat_summary(fun.y = "median", geom = "line", color = "black", size = 1, aes(group = seurat_clusters)) + 
  facet_wrap(.~seurat_clusters, ncol = 3) + 
  scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
  guides(color = "none") + 
  theme_classic() + 
  theme(axis.title = element_blank())
ggsave("./output/Figure2/Figure2E.pdf", height = 8, width = 10)

```

## Figure 2F
```{r}
tfh.28 <- subset(tfh, timepoint == "d28")

circles <- getCirclize(tfh.28, 
                       group.by = "ident")



#Just assigning the normal colors to each cluster
grid.cols <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(tfh$seurat_clusters)))
names(grid.cols) <- levels(tfh@active.ident)


#Graphing the chord diagram
pdf("./output/Figure2/Figure2F_tfh.d28.pdf", height = 4, width = 4)
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = grid.cols, 
                       directional = 1,
                       direction.type =  "arrows",
                       link.arr.type = "big.arrow")
dev.off()

tfh.60 <- subset(tfh, timepoint == "d60")
circles <- getCirclize(tfh.60, 
                       group.by = "ident")

pdf("./output/Figure2/Figure2F_tfh.d60.pdf", height = 4, width = 4)
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = grid.cols, 
                       directional = 1,
                       direction.type =  "arrows",
                       link.arr.type = "big.arrow")
dev.off()

tfh.110 <- subset(tfh, timepoint == "d110")
circles <- getCirclize(tfh.110, 
                       group.by = "ident")

pdf("./output/Figure2/Figure2F_tfh.d110.pdf", height = 4, width = 4)
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = grid.cols, 
                       directional = 1,
                       direction.type =  "arrows",
                       link.arr.type = "big.arrow")
dev.off()


tfh.201 <- subset(tfh, timepoint == "d201")

circles <- getCirclize(tfh.201, 
                       group.by = "ident")


pdf("./output/Figure2/Figure2F_tfh.d201.pdf", height = 4, width = 4)
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = grid.cols, 
                       directional = 1,
                       direction.type =  "arrows",
                       link.arr.type = "big.arrow")
dev.off()

```



## Figure 2G

```{r}
sampled <- df %>%
  group_by(timepoint) %>%
  sample_n(2000)

sampled$timepoint <- factor(sampled$timepoint, levels = c("d28", "d60", "d110", "d201"))
tfh$timepoint <- factor(tfh$timepoint, levels = c("d28", "d60", "d110", "d201"))
df <- data.frame(tfh[[]], tfh@reductions$umap@cell.embeddings)

xmax <- round(max(df$UMAP_1) + 3)
xmin <- round(min(df$UMAP_1) - 3)
ymax <- round(max(df$UMAP_2) + 3)
ymin <- round(min(df$UMAP_2) - 3)

ggplot(sampled, aes(x = UMAP_1, y = UMAP_2)) +
    geom_point(color = "grey") + 
     stat_density_2d(mapping = aes(fill = stat(level)),
                              geom = "polygon", 
                              bins = 20) + 
  geom_point(data = sampled[c(which(sampled$TCRA.S167.180 == "Specific"), which(sampled$TCRB.S167.180 == "Specific")), ], aes(x = UMAP_1, y = UMAP_2), fill = "white", shape = 21, size = 3) + 
          facet_grid(~timepoint) + 
           scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  ylim(ymin, ymax) + 
  xlim(xmin, xmax)+
  guides(fill = "none") + 
  theme_classic()
ggsave("./output/Figure2/Figure2F_TFHonly.pdf", height = 3, width = 10.5)
```


# Figure 3 Graphs

## Loading and Filtering Data

```{r}
dir.create("./output/Figure3/")

list <- readRDS("./data/processedData/Fullintegrated_seuratObjects_TFH.rds")
cells.to.remove <- unique(c(which(nchar(stringr::str_split(list$CTaa, "_", simplify = TRUE)[,1]) > 50), #TRA chain > 50 aa
                     which(nchar(stringr::str_split(list$CTaa, "_", simplify = TRUE)[,2]) > 50))) # TRB chain > 50 aa residues
if(length(cells.to.remove) > 0) {
    cells <- rownames(list[[]])
    cells <- cells[-cells.to.remove]
    list <- subset(list, cells = cells)
}

###################################################
#Generating Seurat Object for individual clonotypes
###################################################

conga <- CoNGAfy(list,
                 meta.carry = c("CTaa", "CTgene"))

meta <- list[[]]
clone.count <- meta %>%
  group_by(CTaa, TCRB.epitope, TCRA.epitope) %>%
  summarise(n = n()) %>%
  as.data.frame()

ID <- clone.count$CTaa
clone.count <- clone.count[,-1]
rownames(clone.count) <- ID

conga <- AddMetaData(conga, clone.count)
saveRDS(conga, "./data/processedData/conga_tfh_seuratObject.rds")

####################
#Adding Meta Data
####################

meta <- list[[]]

CTaa.unique <- unique(meta$CTaa)

for(i in seq_along(CTaa.unique)) {
  meta.tmp <- meta[meta$CTaa == CTaa.unique[i],]
  CTaa.tmp <- c(CTaa = unique(meta.tmp$CTaa), 
                donor = (paste(str_sort(unique(meta.tmp$donor)), collapse = ";")), 
                tissue = paste(str_sort(unique(meta.tmp$tissue)), collapse = ";"), 
                timepoint = paste(str_sort(unique(meta.tmp$timepoint)), collapse = ";"), 
                CD4.annot = paste(str_sort(unique(meta.tmp$CD4.annot)), collapse = ";"), 
                CD8.annot = paste(str_sort(unique(meta.tmp$CD8.annot)), collapse = ";"),
                DICE.annot = paste(str_sort(unique(meta.tmp$DICE.labels)), collapse = ";"), 
                CD8.score = paste(str_sort(unique(meta.tmp$CD8.pos)), collapse = ";"), 
                sequencing.run <- paste(str_sort(unique(meta.tmp$orig.ident)), collapse = ";"), 
                clusters <- paste(str_sort(unique(meta.tmp$seurat_clusters)), collapse = ";"))
  if(i == 1) {
    CTaa.Library <- CTaa.tmp
  } else {
    CTaa.Library <- rbind(CTaa.Library, CTaa.tmp)
  }
}
write.csv(CTaa.Library, "./output/CTaa.tfh.library.csv")

CTaa.Library <- read.csv("./output/CTaa.tfh.library.csv")
CTaa.Library <- CTaa.Library[,-1]
aa <- CTaa.Library$CTaa
CTaa.Library <- CTaa.Library[,-1]
rownames(CTaa.Library) <- aa
colnames(CTaa.Library)[8:9] <- c("SequencingRun", "Orig.Cluster")

conga <- AddMetaData(conga, CTaa.Library)

###################################
#Adding Test TCR sequence information
#####################################

confirming.seq <- list(TCRA = c("CATGLTGGGNKLTF", "CAGQLNRAAGNKLTF", "CIVPRGTGFQKLVF", "CAVRPNAGNNRKLIW", "CAVTGTYKYIF"), 
                       TCRB = c("CASSMGGGNQPQHF", "CASSPRGTEAFF", "CASSLVGGGPAEAFF", "CASSLGENDEQFF", "CASSQGNRANTEAFF")) 

conga$confirmation.seq <- NA
for (i in seq_len(5)) {
  x.pos <- intersect(grep(confirming.seq[[1]][i], conga$CTaa), grep(confirming.seq[[2]][i], conga$CTaa))
  conga$confirmation.seq[x.pos] <- paste0("TCR.", i)
}

conga$TRAV <- str_split(conga$CTgene, "[.]", simplify = TRUE)[,1]
conga$TRAV <- factor(conga$TRAV, levels = str_sort(unique(conga$TRAV), numeric = TRUE))
conga$TRBV <- str_split(str_split(conga$CTgene, "_", simplify = TRUE)[,2], "[.]", simplify = TRUE)[,1]
conga$TRBV <- factor(conga$TRBV, levels = str_sort(unique(conga$TRBV), numeric = TRUE))


#################################
#Performing Dimensional Reduction
#################################
dimen <- 15

conga <- conga %>% 
            NormalizeData(verbose = FALSE) %>% 
            FindVariableFeatures() 
conga <- Trex::quietTCRgenes(conga)
conga <- ScaleData(conga, verbose = FALSE) 
conga <- RunPCA(conga, features = VariableFeatures(conga))

conga <- runTrex(conga, 
                 chains = "TRA", 
                 AA.properties = "KF", 
                 reduction.name = "TRA.KF")

conga <- runTrex(conga, 
                 chains = "TRB", 
                 AA.properties = "KF", 
                 reduction.name = "TRB.KF")


dem.tmp <- c(1:dimen, 31:(31+dimen-1), 61:(61+dimen-1))

#Calculating mutual neighbor values
output <- rescaleByNeighbors(
                  list(conga[["TRA.KF"]]@cell.embeddings[,1:dimen], 
                      conga[["TRB.KF"]]@cell.embeddings[,1:dimen],
                      conga[["pca"]]@cell.embeddings[,1:dimen]), 
                  k = 100)

conga[["KF.corrected"]] <- CreateDimReducObject(
                                embeddings = as.matrix(output),
                                stdev = rep(0, ncol(output)),
                                key = "correctedTrex",
                                jackstraw = NULL,
                                misc = list())

#Calculating PHATE embedding
phate.results <- phate(conga[["KF.corrected"]]@cell.embeddings[,1:length(dem.tmp)])


        
        
conga[["phateTrex"]] <- CreateDimReducObject(
                                        embeddings = phate.results$embedding,
                                        stdev = rep(0, 2),
                                        key = "phateTrex",
                                        jackstraw = NULL,
                                        misc = list())

#Calculating Neighbors
conga <- FindNeighbors(conga, 
                         reduction = "KF.corrected",
                         dims = 1:length(dem.tmp),
                         force.recalc = TRUE,
                         graph.name = c("T_nn", "T_snn"))

conga <- FindClusters(conga, algorithm = 3, graph.name = "T_snn", resolution = 0.3)

saveRDS(conga, "./data/processedData/conga_tfh_seuratObject.rds")
```


## Figure 3B
```{r}
conga <- readRDS("./data/processedData/conga_tfh_seuratObject.rds")
df <- data.frame(conga@reductions$phateTrex@cell.embeddings,
                  conga[[]])

centers = df %>% 
          group_by(seurat_clusters) %>% 
          summarize(across(c("phateTrex_1", "phateTrex_2"),mean)) %>%
          as.data.frame()

max.cluster.val <- max(as.numeric(na.omit(df$seurat_clusters)))

mycolors <- colorRampPalette(brewer.pal(12,"Paired"))(length(unique(df$seurat_clusters)))

ggplot() +
    geom_point(data =df, aes(x=phateTrex_1, y = phateTrex_2, size = n, fill = seurat_clusters), 
               shape =21, stroke = 0.1) + 
    scale_fill_manual(values = mycolors) + 
    theme_classic() +
    geom_text(data = centers, mapping =  aes(x=centers[,2], y = centers[,3], label = seurat_clusters), 
              color = "black", size = 5) +
    labs(fill = "cluster") + 
    theme(axis.title = element_blank())

ggsave("./output/Figure3/Figure3B.png", height = 4, width = 5.75, dpi = 600)
```

## Figure 3C
```{r}
x.features <- c("BCL6", "CXCL13", "CXCR5", "ICOS", "MAF", "PDCD1", "EMP3", "KLF2")
x <- FeaturePlot(conga, 
              features = x.features,
              reduction = "phateTrex",
            cols = rev(brewer.pal(11, "RdYlBu")), 
            coord.fixed = FALSE, 
            combine = FALSE) 
for(i in seq_along(x)) {
  x[[i]] <- x[[i]] + 
    xlim(-0.03,0.05) +
    ylim(-0.01,0.015) +
    guides(color = "none") + 
    theme_void()
}
plot_a_list(x,4,2)
ggsave("./output/Figure3/Figure3C.png", height = 10.67, width = 6.67, dpi = 600)
```

## Figure 3D
```{r}
epitope <- stringr::str_sort(na.omit(unique(df$TCRA.epitope)), numeric = TRUE)
        
ep.colors <- rev(brewer.pal(5, "RdYlBu"))[c(1,3,5)]
names(ep.colors) <- epitope

df$TCRA.epitope <- factor(df$TCRA.epitope, levels = epitope)
  
# Plotting the TCRA-specific sequences
df.sub <- subset(df, !is.na(TCRA.epitope))
ggplot() +
  geom_point(data =df, aes(x=phateTrex_1, y = phateTrex_2, size = n), 
             fill = "grey", shape =21, stroke = 0.1) + 
  geom_point(data = df.sub, aes(x=phateTrex_1, y = phateTrex_2, size = n, fill = TCRA.epitope), 
             shape =21, stroke = 0.1) + 
  scale_fill_manual(values = ep.colors) + 
  guides(size = "none", fill = "none") + 
  theme_void()
  ggsave("./output/Figure3/Figure3D_TRA.png", height = 4, width = 5, dpi = 600)

# Plotting the TCRB-specific sequences
df$TCRB.epitope <- factor(df$TCRB.epitope, levels = epitope)

df.sub <- subset(df, !is.na(TCRB.epitope))
ggplot() +
    geom_point(data =df, aes(x=phateTrex_1, y = phateTrex_2, size = n), 
               fill = "grey", shape =21, stroke = 0.1) + 
    geom_point(data = df.sub, aes(x=phateTrex_1, y = phateTrex_2, size = n, fill = TCRB.epitope), 
               shape =21, stroke = 0.1) + 
    scale_fill_manual(values = ep.colors) + 
    guides(size = "none", fill = "none") + 
    theme_void()

ggsave("./output/Figure3/Figure3D_TRB.png", height = 4, width = 5, dpi = 600)
```

## Figure 3E

```{r}

clusters <- c(0,1,2)
conga.meta <- conga[[]]
for(i in seq_along(clusters)) {
  tmp.meta <- conga.meta[conga.meta$seurat_clusters == clusters[i],]
  tmp.meta <- tmp.meta[!is.na(tmp.meta$TCRB.epitope) |!is.na(tmp.meta$TCRA.epitope),] 
  TRB <- str_split(tmp.meta$CTaa, "_", simplify = TRUE)[,2]
  TRA <- str_split(tmp.meta$CTaa, "_", simplify = TRUE)[,1]
  align <- msa(TRA, 
               type = "protein", 
               maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave(paste0("./output/Figure3/Figure3E.Cluster.", clusters[i], ".TRA.pdf"), height =3, width = 4.5)
  
    align <- msa(TRB, 
               type = "protein", 
               maxiters = 30)
  align2 <- msaConvert(align, type="seqinr::alignment")
  
  ggseqlogo(align2$seq, 
            seq_type='aa', 
            method='p') 
  ggsave(paste0("./output/Figure3/Figure3E.Cluster.", clusters[i], ".TRB.pdf"), height =3, width = 4.5)
}
```


## Figure 3F
```{r}
df.sub <- subset(df, !is.na(confirmation.seq))

ggplot() +
    geom_point(data =df, aes(x=phateTrex_1, y = phateTrex_2, size = n), 
               fill = "grey", shape =21, stroke = 0.1) + 
    geom_point(data = df.sub, aes(x=phateTrex_1, y = phateTrex_2, fill = confirmation.seq), 
               size = 4, shape =21, stroke = 0.2) + 
    scale_fill_manual(values = brewer.pal(5, "RdYlBu")) +
    theme_void() + 
    guides(fill = "none", size = "none")

ggsave("./output/Figure3/Figure3F.png", height = 4, width = 5, dpi = 600)
```

# Figure 4 Graphs

```{r cars}

conga <- readRDS("./data/processedData/conga_tfh_seuratObject.rds")
conga <- annotateDB(conga, chains = "TRB", edit.distance = 0)
conga <- annotateDB(conga, chains = "TRA", edit.distance = 0)

meta <- conga[[]]
meta <- meta[!is.na(meta$confirmation.seq),]
meta <- meta[order(meta$confirmation.seq),]

confirm.TCRA.gene <- as.vector(meta$TRAV)
confirm.TCRB.gene <- as.vector(meta$TRBV)
confirm.TCRA <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,1]
confirm.TCRB <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,2]
edit.distance <- 2

TCRA <- stringr::str_split(conga[[]]$CTaa, "_", simplify = TRUE)[,1]
TCRB <- stringr::str_split(conga[[]]$CTaa, "_", simplify = TRUE)[,2]
    
additions.TCRA <- lapply(confirm.TCRA, function(x) {
      pos <- which(stringdist(x, TCRA, method = "lv") <= edit.distance & 
                     stringdist(x, TCRA, method = "lv") > 0)
      pos
})

additions.TCRB <- lapply(confirm.TCRB, function(x) {
      pos <- which(stringdist(x, TCRB, method = "lv") <= edit.distance & 
                     stringdist(x, TCRB, method = "lv") > 0)
      pos
    })

conga@meta.data$confirmation.neighbors.TCRA <- NA
conga@meta.data$confirmation.neighbors.TCRB <- NA

for(i in seq_along(additions.TCRA)) {
  idx.A <- as.vector(conga@meta.data$TRAV[additions.TCRA[[i]]])
  idx.A <- which(idx.A == confirm.TCRA.gene[i])
  
  conga@meta.data$confirmation.neighbors.TCRA[additions.TCRA[[i]][idx.A]] <- paste0("TCR.", i)
}

for(i in seq_along(additions.TCRB)) {
   idx.B <- as.vector(conga@meta.data$TRBV[additions.TCRB[[i]]])
  idx.B <- which(idx.B == confirm.TCRB.gene[i])
  conga@meta.data$confirmation.neighbors.TCRB[additions.TCRB[[i]][idx.B]] <- paste0("TCR.", i)
}

meta <- conga[[]]
conga@meta.data$TCR.both <- paste0(conga@meta.data$confirmation.neighbors.TCRA, conga@meta.data$confirmation.neighbors.TCRB)
conga@meta.data$TCR.both[conga@meta.data$TCR.both == "NANA"] <- NA
conga@meta.data$TCR.both <- stringr::str_remove_all(conga@meta.data$TCR.both, "NA")
conga@meta.data$TCR.both[conga@meta.data$TCR.both == "TCR.1TCR.1"] <- "TCR.1"
conga@meta.data$TCR.both[conga@meta.data$TCR.both == "TCR.2TCR.2"] <- "TCR.2"

df <- data.frame(conga@reductions$phateTrex@cell.embeddings,
                  conga[[]])


df.sub <- subset(df, !is.na(confirmation.seq))
df.sub2 <- subset(df, !is.na(TCR.both))
write.csv(df.sub2, "Trex_edit.distance2_matchedVgenes.csv")



saveRDS(conga, "./data/processedData/conga_tfh_seuratObject.rds")

dir.create("./output/Figure4")

conga <- readRDS("./data/processedData/conga_tfh_seuratObject.rds")
conga$TCRA.COVID <- "No"
conga$TCRA.COVID[grepl("Spike", conga$TRA_Epitope.target)] <- "Yes"
conga$TCRA.COVID[!is.na(conga$TCRA.epitope)] <- "Yes"
conga$TCRA.COVID[!is.na(conga@meta.data$confirmation.neighbors.TCRA)] <- "Yes"
TCRA.ref <- data.frame("CTaa" = conga$CTaa, "COVID" = conga$TCRA.COVID)
TCRA.ref <- subset(TCRA.ref, COVID == "Yes")

conga$TCRB.COVID <- "No"
conga$TCRB.COVID[grepl("Spike", conga$TRB_Epitope.target)] <- "Yes"
conga$TCRB.COVID[!is.na(conga$TCRB.epitope)] <- "Yes"
conga$TCRB.COVID[!is.na(conga@meta.data$confirmation.neighbors.TCRB)] <- "Yes"
TCRB.ref <- data.frame("CTaa" = conga$CTaa, "COVID" = conga$TCRB.COVID)
TCRB.ref <- subset(TCRB.ref, COVID == "Yes")



SeuratObj <- readRDS("./data/processedData/Fullintegrated_seuratObjects_TFH_TFHonly.rds")

updated.shared.clones <- sort(unique(c(TCRB.ref$CTaa, TCRA.ref$CTaa)))
write.csv(updated.shared.clones, "./output/Figure4/Figure4_spike.clones.csv")
```

## Figure 4A

```{r}
SeuratObj <- subset(SeuratObj, timepoint != "d110")
subset.obj <- subset(SeuratObj, CTaa %in% updated.shared.clones)

Idents(subset.obj) <- "timepoint"
FvL.markers <- FindMarkers(subset.obj, 
                           ident.1 = "d28",
                           ident.2 = "d201",
                           test.use = "MAST", 
                           latent.vars = "donor", 
                           logfc.threshold = 0,
                           min.pct = 0,
                           pseudocount.use = 0.1)

write.csv(FvL.markers, "./output/Figure4/Figure4A_Tfh.DEG.d28vsd201.csv")

FvL.markers <- FvL.markers %>%
                mutate(Percent.Delta = pct.1 - pct.2)
FvL.markers$genes <- rownames(FvL.markers)

top10 <- FvL.markers %>% 
  subset(p_val < 0.05) %>%
  slice_max(order_by = avg_log2FC, n = 10)

bottom10 <- FvL.markers %>% 
  subset(p_val < 0.05) %>%
  slice_min(order_by = avg_log2FC, n = 10)

rb.palette <- colorRampPalette(brewer.pal(11, "RdYlBu"))(11)
FvL.plot <- ggplot(FvL.markers, aes(x = avg_log2FC, y = -log10(p_val))) + 
  geom_point(aes(size = Percent.Delta), color = "grey") + 
  geom_point(data=subset(FvL.markers, genes %in% top10$genes),
             aes(size = Percent.Delta), color = rb.palette[1]) + 
  geom_point(data=subset(FvL.markers, genes %in% bottom10$genes),
             aes(size = Percent.Delta), color = rb.palette[11]) + 
  geom_text_repel(data=subset(FvL.markers, genes %in% c(top10$genes,bottom10$genes)), aes(label = genes), max.overlaps = 20) + 
  theme_classic() + 
  scale_size(range = c(0.1,3)) + 
  geom_hline(yintercept = 1.3) + 
  geom_vline(xintercept = 0)

FvL.plot
ggsave("./output/Figure4/Figure4A.png", height = 4.5, width = 5, dpi = 600)

```

## Figure 4B

```{r}
ES <- readRDS("./data/processedData/enrichmentResults.rds")
meta <- subset.obj[[]]

ES2 <- ES[rownames(ES) %in% rownames(meta),]
colnames(ES2) <- stringr::str_remove_all(colnames(ES2), "_UCell")
variances <- apply(ES2, 2, var)
#ES2 <- ES2[,-which(variances == 0)]

pathways <- colnames(ES2)

ES2 <- merge(ES2, meta, by = 0)

sig.results <- getSignificance(ES2, 
                               group = "timepoint", 
                               gene.sets = pathways, 
                               fit = "KW")
rownames(sig.results) <- stringr::str_remove_all(rownames(sig.results), "_UCell")

sig.results2 <- getSignificance(ES2, 
                               group = "timepoint", 
                               gene.sets = pathways, 
                               fit = "ANOVA")
rownames(sig.results2) <- stringr::str_remove_all(rownames(sig.results2), "_UCell")

KW.sig.pathways <- sig.results %>%
                        filter(p.value < 0.05) %>%
                        slice_max(order_by = Chi.square, n = 400) %>%
                        rownames() 

ANOVA.sig.pathways <- sig.results2 %>%
                        filter(p.value < 0.05) %>%
                        slice_max(order_by = f.value, n = 400) %>%
                        rownames()

pathways.intersect <- intersect(KW.sig.pathways, ANOVA.sig.pathways)

pathways.intersect <- pathways.intersect[!grepl("REACTOME", pathways.intersect) ]
pathways.intersect <- pathways.intersect[-which(pathways.intersect %in% c( "BIOCARTA_FIBRINOLYSIS_PATHWAY", "BIOCARTA_AMI_PATHWAY", "BIOCARTA_EXTRINSIC_PATHWAY"))]


write(pathways.intersect, "output/Figure4/Figure4B_intersected.pathways.txt")

summary <- ES2 %>%
  group_by(timepoint) %>%
  summarise(across(pathways.intersect, median))

normalize <- function(x) {
      (x- min(x)) /(max(x)-min(x))
}

for(i in 2:74) {
  summary[,i] <- normalize(summary[,i])
}
summary <- as.data.frame(summary)

rownames(summary) <- summary$timepoint



pdf("./output/Figure4/Figure4b.pdf", height = 10, width = 10)
pheatmap::pheatmap(t(summary[,2:74]), scale = "row")
dev.off()
```

## Figure 4C 

```{r}
"%!in%" <- Negate("%in%")


SeuratObj <- readRDS("./data/processedData/Fullintegrated_seuratObjects_TFH_TFHonly.rds")
SeuratObj <- subset(SeuratObj, timepoint != "d110")

df <- unique(SeuratObj[[]][SeuratObj$CTaa %in% updated.shared.clones, c("CTaa", "timepoint")])
duplicated <- unique(df$CTaa[which(duplicated(df$CTaa))])
not.duplicated <- unique(df$CTaa[which(!duplicated(df$CTaa))])
not.duplicated <- not.duplicated[not.duplicated %!in% duplicated]


clone.palette <- colorRampPalette(brewer.pal(11, "RdYlBu"))(length(duplicated))

names(clone.palette) <- duplicated
pal2 <- rep("grey", length(not.duplicated))
names(pal2) <- not.duplicated
clone.palette <- c(clone.palette, pal2)

return.cc <- compareClonotypes(SeuratObj, 
                  clonotypes = unique(df$CTaa), 
                  split.by = "timepoint",
                  cloneCall="aa", 
                  graph = "alluvial",
                  exportTable = TRUE)

all.times <- c("d28", "d60", "d201")
output <- NULL
for(i in seq_along(duplicated)) {
  x <- return.cc[return.cc$Clonotypes == duplicated[i],]
  placement <- names(which(table(x$Sample) == 1))
  new.times <- all.times[all.times %!in% placement]
  rep.value <- length(new.times)
  y <- data.frame(Clonotypes = rep(duplicated[i], rep.value), Proportion = rep(0, rep.value), Sample = new.times)
  output <- rbind(output, y)
  
}
  return.cc <- unique(rbind(return.cc, output))
  return.cc$Sample <- factor(return.cc$Sample, levels = c("d28", "d60", "d201"))
  
  return.cc$Clonotypes <- factor(return.cc$Clonotypes, levels = names(clone.palette))
  ggplot(return.cc, aes(x = Sample, 
                        fill = Clonotypes, 
                        group = Clonotypes,
                        stratum = Clonotypes, 
                        alluvium = Clonotypes, 
                         y = Proportion, 
                        label = Clonotypes)) +
                theme_classic() +
                theme(axis.title.x = element_blank()) + 
                geom_stratum() +
                geom_flow(stat = "alluvium") + 
                guides(fill = "none")  + 
                scale_fill_manual(values = clone.palette)
  ggsave("./output/Figure4/Figure4C.pdf", height = 4, width = 6)
```

# Figure 5 Graphs

```{r}
dir.create("./output/Figure5")

list <- readRDS("./data/processedData/Fullintegrated_seuratObjects_Tcells.rds")

list <- subset(list, donor %in% c("368-01a", "368-13", "368-22"))

meta <- list[[]]
meta$TRA <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,1]
meta$TRB <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,2]
```


## Figure 5A

```{r}

dat.output <- NULL
subjects <- c("368-01a", "368-13", "368-22")
count <- 1

meta2 <- meta

for(i in seq_along(subjects)) {
  
  meta.tmp <- meta2[meta2$donor == subjects[i],]
  meta.tmp.B <- meta.tmp[meta.tmp$tissue == "PB",]
  meta.tmp.L <- meta.tmp[meta.tmp$tissue != "PB",]
  b.counts <- unname(table(meta.tmp.B$CTaa))
  l.counts <- unname(table(meta.tmp.L$CTaa))
  dat.output[[count]] <- b.counts
  dat.output[[(count + 1)]] <- l.counts
  count <- (i*2)+1
}

names(dat.output) <- c("368-01a_B", "368-01a_LN", "368-13_B", "368-13_LN", "368-22_B", "368-22_LN")

out <- iNEXT(dat.output, q=0, datatype="abundance") 

ggiNEXT(out, type=2) 
ggsave("./output/Figure5/Figure5A.pdf", height = 7, width = 7)
```

## Figure 5B

```{r}
subjects <- c("368-01a", "368-13", "368-22")

dat.output <- NULL
for(i in seq_along(subjects)) {
  meta.tmp <- meta[meta$donor == subjects[i],]
  meta.tmp.B <- meta.tmp[meta.tmp$tissue == "PB",]
  meta.tmp.L <- meta.tmp[meta.tmp$tissue != "PB",]
  
  TRA.intersect <- length(intersect(unique(meta.tmp.B$TRA), unique(meta.tmp.L$TRA)))
  TRA.union <- length(unique(meta.tmp.B$TRA, meta.tmp.L$TRA))
  TRA.blood <- length(unique(meta.tmp.B$TRA))
  TRA.LN <- length(unique(meta.tmp.L$TRA))
                                
  
  TRB.intersect <- length(intersect(unique(meta.tmp.B$TRB), unique(meta.tmp.L$TRB)))
  TRB.union <- length(unique(meta.tmp.B$TRB, meta.tmp.L$TRB))
  TRB.blood <- length(unique(meta.tmp.B$TRB))
  TRB.LN <- length(unique(meta.tmp.L$TRB))
  
  combined.intersect <- length(intersect(unique(meta.tmp.B$CTaa), unique(meta.tmp.L$CTaa)))
  combined.union <- length(unique(meta.tmp.B$CTaa, meta.tmp.L$CTaa))
  combined.blood <- length(unique(meta.tmp.B$CTaa))
  combined.LN <- length(unique(meta.tmp.L$CTaa))
  
  out <- c(subjects[i], TRA.intersect, TRA.union, TRA.blood, TRA.LN, 
           TRB.intersect, TRB.union, TRB.blood, TRB.LN, 
           combined.intersect, combined.union, combined.blood, combined.LN)
  dat.output <- rbind(dat.output, out)
}
colnames(dat.output) <- c("subjects", "TRA.intersect", "TRA.union", "TRA.blood", "TRA.LN", 
                          "TRB.intersect", "TRB.union", "TRB.blood", "TRB.LN",
                          "combined.intersect", "combined.union", "combined.blood", "combined.LN")
write.csv(dat.output, "./output/Figure5/Figure5_TCRsummary.csv")

chain.combo <- c("TRA", "TRB", "combined") 
for (i in seq_along(subjects)) {
  tmp <- dat.output[i,]
  for(j in seq_along(chain.combo)) {
    chain.tmp <- c(BL = as.numeric(tmp[(4*j)]), LN = as.numeric(tmp[(4*j)+1]), "BL&LN" = as.numeric(tmp[(4*j)-2]))
    fit <- euler(chain.tmp, shape = "ellipse")
    venn.plot <- plot(fit, 
         quantities = TRUE,
         labels = list(font = 4))
    pdf(paste0("./output/Figure5/Figure5B_", subjects[i], ".", chain.combo[j], "_allTCRs.pdf"), height = 3, width = 3.5)
      print(venn.plot)
    dev.off()
  }
}
```

## Figure 5C

```{r}

list$compare <- paste0(list$donor, "_", list$tissue)
plot1 <- scatterClonotype(list, 
                 split.by = "compare",
                 cloneCall ="aa", 
                 x.axis = "368-01a_LN", 
                 y.axis = "368-01a_PB") + 
  scale_color_manual(values = rev(brewer.pal(5, "RdYlBu"))[c(1,2,5,4,3)])

plot2 <- scatterClonotype(list, 
                 split.by = "compare",
                 cloneCall ="aa", 
                 x.axis = "368-22_LN", 
                 y.axis = "368-22_PB",
                 dot.size = "total") + 
  scale_color_manual(values = rev(brewer.pal(5, "RdYlBu"))[c(1,2,5,4,3)])


plot1 <- plot1 + guides(color = "none", size = "none") + theme(axis.title = element_blank())
plot2 <- plot2 + guides(color = "none", size = "none") + theme(axis.title = element_blank())

plot1 + plot2 + plot_layout(ncol = 1) 
ggsave("./output/Figure5/Figure5C.pdf", height = 8, width = 4)
```

# Figure 6 Graphs

```{r}
dir.create("./output/Figure6")
seuratObj <- readRDS("./data/processedData/Fullintegrated_seuratObjects_Tcells.rds")
seuratObj <- annotateDB(seuratObj, chains = "TRB", edit.distance = 0)
seuratObj <- annotateDB(seuratObj, chains = "TRA", edit.distance = 0)

seuratObj$TRAV <- stringr::str_split(seuratObj$CTgene, "[.]", simplify = TRUE)[,1]
seuratObj$TRBV <- stringr::str_split(seuratObj$CTgene, "_", simplify = TRUE)[,2]
seuratObj$TRBV <- stringr::str_split(seuratObj$TRBV, "[.]", simplify = TRUE)[,1]

meta <- conga[[]]
meta <- meta[!is.na(meta$confirmation.seq),]
meta <- meta[order(meta$confirmation.seq),]
confirm.TCRA.gene <- as.vector(meta$TRAV)
confirm.TCRB.gene <- as.vector(meta$TRBV)
confirm.TCRA <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,1]
confirm.TCRB <- stringr::str_split(meta$CTaa, "_", simplify = TRUE)[,2]
edit.distance <- 2

TCRA <- stringr::str_split(seuratObj[[]]$CTaa, "_", simplify = TRUE)[,1]
TCRB <- stringr::str_split(seuratObj[[]]$CTaa, "_", simplify = TRUE)[,2]
    
additions.TCRA <- lapply(confirm.TCRA, function(x) {
      pos <- which(stringdist(x, TCRA, method = "lv") <= edit.distance & 
                     stringdist(x, TCRA, method = "lv") > 0)
      pos
})

additions.TCRB <- lapply(confirm.TCRB, function(x) {
      pos <- which(stringdist(x, TCRB, method = "lv") <= edit.distance & 
                     stringdist(x, TCRB, method = "lv") > 0)
      pos
    })

seuratObj@meta.data$confirmation.neighbors.TCRA <- NA
seuratObj@meta.data$confirmation.neighbors.TCRB <- NA

for(i in seq_along(additions.TCRA)) {
  idx.A <- as.vector(seuratObj@meta.data$TRAV[additions.TCRA[[i]]])
  idx.A <- which(idx.A == confirm.TCRA.gene[i])
  
  seuratObj@meta.data$confirmation.neighbors.TCRA[additions.TCRA[[i]][idx.A]] <- paste0("TCR.", i)
}

for(i in seq_along(additions.TCRB)) {
   idx.B <- as.vector(seuratObj@meta.data$TRBV[additions.TCRB[[i]]])
  idx.B <- which(idx.B == confirm.TCRB.gene[i])
  seuratObj@meta.data$confirmation.neighbors.TCRB[additions.TCRB[[i]][idx.B]] <- paste0("TCR.", i)
}

#Spike 167-180 epitope
sequence <- read.csv("./sequenceInfo/Sequences.csv")
CTaa <- as.data.frame(stringr::str_split(seuratObj$CTaa, "_", simplify = TRUE))
TA <- which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

seuratObj$TCRB.S167.180<- NA
seuratObj$TCRB.S167.180[TB] <- "Specific"

seuratObj$TCRA.S167.180 <- NA
seuratObj$TCRA.S167.180[TA] <- "Specific"

#Spike 815-830 epitope

sequence <- read.csv("./sequenceInfo/dykema.sequences.csv")
CTaa <- as.data.frame(stringr::str_split(seuratObj$CTaa, "_", simplify = TRUE))
TA <-which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

seuratObj$TCRB.S815.830 <- NA
seuratObj$TCRB.S815.830[TB] <- "Specific"

seuratObj$TCRA.S815.830 <- NA
seuratObj$TCRA.S815.830[TA] <- "Specific"

# VDJdb sequence

sequences <- read.delim("./sequenceInfo/VDJdb.tsv")
sequences <- sequences[sequences$MHC.class == "MHCII" & sequences$Epitope.gene == "Spike",]

epitope <- c("NCTFEYVSQPFLMDL" = "S164.179",
"RISNCVADYSVLYNS" = "S356.371",
"NFSQILPDPSKPSKR" = "S800.815",
"HWFVTQRNFYEPQII" = "S1101.1115",
"VGGNYNYLYRLFRKS" = "S445.459",
"NNSYECDIPIGAGIC" = "S657.671")

seuratObj$TCRB.VDJ.db <- NA
seuratObj$TCRA.VDJ.db <- NA
for (i in seq_along(epitope)) {
  sequences$Epitope[which(sequences$Epitope == names(epitope)[i])] <- epitope[i]
  TA <- sequences[sequences$Gene == "TRA" & sequences$Epitope == epitope[i], ]
  TB <- sequences[sequences$Gene == "TRB" & sequences$Epitope == epitope[i], ]
  TA.loc <-which(CTaa$V1 %in% TA$CDR3)
  TB.loc <-which(CTaa$V2 %in% TB$CDR3)
  seuratObj$TCRB.VDJ.db[TB.loc] <- epitope[i]
  seuratObj$TCRA.VDJ.db[TA.loc] <- epitope[i]
}

seuratObj$TCRB.epitope  <- NA
seuratObj$TCRB.epitope <- ifelse(!is.na(seuratObj$TCRB.S167.180), "S167.180", NA)
seuratObj$TCRB.epitope <- ifelse(!is.na(seuratObj$TCRB.S815.830), "S815.830", seuratObj$TCRB.epitope)
seuratObj$TCRB.epitope <- ifelse(!is.na(seuratObj$TCRB.VDJ.db) & is.na(seuratObj$TCRB.epitope), seuratObj$TCRB.VDJ.db, seuratObj$TCRB.epitope)

seuratObj$TCRA.epitope  <- NA
seuratObj$TCRA.epitope <- ifelse(!is.na(seuratObj$TCRA.S167.180), "S167.180", NA)
seuratObj$TCRA.epitope <- ifelse(!is.na(seuratObj$TCRA.S815.830), "S815.830", seuratObj$TCRA.epitope)
seuratObj$TCRA.epitope <- ifelse(!is.na(seuratObj$TCRA.VDJ.db) & is.na(seuratObj$TCRA.epitope), seuratObj$TCRA.VDJ.db, seuratObj$TCRA.epitope)

seuratObj$TCRA.COVID <- "No"
seuratObj$TCRA.COVID[grepl("Spike", seuratObj$TRA_Epitope.target)] <- "Yes"
seuratObj$TCRA.COVID[!is.na(seuratObj$TCRA.epitope)] <- "Yes"
seuratObj$TCRA.COVID[!is.na(seuratObj$confirmation.neighbors.TCRA)] <- "Yes"
TCRA.ref <- data.frame("CTaa" = seuratObj$CTaa, "COVID" = seuratObj$TCRA.COVID)
TCRA.ref <- subset(TCRA.ref, COVID == "Yes")

seuratObj$TCRB.COVID <- "No"
seuratObj$TCRB.COVID[grepl("Spike", seuratObj$TRB_Epitope.target)] <- "Yes"
#seuratObj$TCRB.COVID[!is.na(seuratObj$TCRB.epitope)] <- "Yes"
seuratObj$TCRB.COVID[!is.na(seuratObj$confirmation.neighbors.TCRB)] <- "Yes"
TCRB.ref <- data.frame("CTaa" = seuratObj$CTaa, "COVID" = seuratObj$TCRB.COVID)
TCRB.ref <- subset(TCRB.ref, COVID == "Yes")

write.csv(unique(c(TCRA.ref$CTaa, TCRB.ref$CTaa)), "./output/Figure6/Figure6_spike.clones.csv")

subset.Obj <- subset(seuratObj, CTaa %in% unique(c(TCRA.ref$CTaa, TCRB.ref$CTaa)))
subset.Obj <- subset(subset.Obj, CD8.pos == "No")

```

# Figure 6A

```{r}
Idents(subset.Obj) <- "tissue"
LNvB.markers <- FindMarkers(subset.Obj, 
                           ident.1 = "LN",
                           test.use = "MAST", 
                           latent.vars = c("donor"), 
                           logfc.threshold = 0,
                           min.pct = 0,
                           pseudocount.use = 0.1)

write.csv(LNvB.markers, "./output/Figure6/Figure6A_Ag.specific.LNvsBL.csv")
LNvB.markers <- LNvB.markers %>%
                mutate(Percent.Delta = pct.1 - pct.2)
LNvB.markers$genes <- rownames(LNvB.markers)


top10 <- LNvB.markers %>% 
  subset(p_val_adj < 0.05) %>%
  slice_max(order_by = avg_log2FC, n = 10)

bottom10 <- LNvB.markers %>% 
  subset(p_val_adj < 0.05) %>%
  slice_min(order_by = avg_log2FC, n = 10)

LNvBL.plot <- ggplot(LNvB.markers, aes(x = avg_log2FC, y = -log10(p_val_adj))) + 
  geom_point(aes(size = Percent.Delta), color = "grey") + 
  geom_point(data=subset(LNvB.markers, genes %in% top10$genes),
             aes(size = Percent.Delta), color = rb.palette[1]) + 
  geom_point(data=subset(LNvB.markers, genes %in% bottom10$genes),
             aes(size = Percent.Delta), color = rb.palette[11]) + 
  geom_text_repel(data=subset(LNvB.markers, genes %in% c(top10$genes,bottom10$genes)), aes(label = genes), max.overlaps = 30) + 
  theme_classic() + 
  scale_size(range = c(0.1,3)) + 
  geom_hline(yintercept = 1.3) + 
  geom_vline(xintercept = 0)
ggsave("./output/Figure6/Figure6A.pdf", height = 4.5, width = 5, dpi = 600)
```

## Figure 6B

```{r}
meta <- subset.Obj[[]]

ES2 <- ES[rownames(ES) %in% rownames(meta),]
colnames(ES2) <- stringr::str_remove_all(colnames(ES2), "_UCell")
variances <- apply(ES2, 2, var)
#ES2 <- ES2[,-which(variances == 0)]

pathways <- colnames(ES2)

ES2 <- merge(ES2, meta, by = 0)

sig.results <- getSignificance(ES2, 
                               group = "tissue", 
                               gene.sets = pathways, 
                               fit = "Wilcoxon")

sig.results2 <- getSignificance(ES2, 
                               group = "tissue", 
                               gene.sets = pathways, 
                               fit = "T.test")

W.sig.pathways <- sig.results %>%
                        filter(p.value < 0.05) %>%
                        slice_max(order_by = W.statistic, n = 200) %>%
                        rownames() 

T.sig.pathways <- sig.results2 %>%
                        filter(p.value < 0.05) %>%
                        slice_max(order_by = T.statistic, n = 200) %>%
                        rownames()

pathways.intersect <- intersect(W.sig.pathways, T.sig.pathways)

pathways.intersect <- pathways.intersect[!grepl("REACTOME", pathways.intersect) ]
write(pathways.intersect, "./output/Figure6/Figure6B_Agspecific.intersected.pathways.txt")

ES2$timetissue <- paste0(ES2$timepoint, "_", ES2$tissue)
summary <- ES2 %>%
  group_by(timetissue) %>%
  summarise(across(pathways.intersect, median))

summary <- as.data.frame(summary)
rownames(summary) <- summary$timetissue

pdf("./output/Figure6/Figure6b.pdf", height = 8, width = 14)
pheatmap::pheatmap(t(summary[,2:48]), scale = "row")
dev.off()
```

# Figure 7 Graphs

```{r, eval= FALSE}
inventory <- read.delim("data.inventory.txt")
inventory <- inventory[inventory$tissue == "PB",]
samples<- inventory$sample

list <- list()
for (i in seq_along(samples)) {
  list[[i]] <- readRDS(paste0("./data/", samples[i], "/processedRNA_SeuratObject.rds"))
}

#Merging Data
list <- merge(list[[1]], list[-1])

meta <- list[[]]
#meta <- meta[,-which(colnames(meta) %in% colnames(data.inventory))]
barcodes <- rownames(meta)
meta <- merge(meta, inventory, by.x = "orig.ident", by.y = "sample")
meta <- meta[,colnames(inventory)[-2]]
rownames(meta) <- barcodes

list <- AddMetaData(list, meta)


list <- list %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(verbose = FALSE) %>% 
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident")) %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:25,
            n.epochs = 500) %>%
    FindClusters(resolution = 1, 
                 algorithm = 3) %>% 
    identity()

saveRDS(list, file = "./data/processedData/PB_Fullintegrated_seuratObjects.rds")

library(scRepertoire)

######################################
#iterate to make a list of contig csvs
######################################


contig.list <- NULL
for (i in seq_along(samples)){
  contig.list[[i]] <-  read.csv(paste0("./data/", samples[i], "/filtered_contig_annotations.csv"))
}
names(contig.list) <- samples

##################################################
#Reducing the data to the individual barcode level
##################################################
combinedObject <- combineTCR(contig.list, samples = samples, filterMulti = TRUE, removeNA = TRUE, cells = "T-AB")

pt <- inventory$donor[match(samples, inventory$sample)]
combinedObject <- addVariable(combinedObject, name = "Patient", variables = pt)
list <- combineExpression(combinedObject, list, cloneCall = "aa", group.by = "Patient")

slot(list, "meta.data")$cloneType <- factor(slot(list, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (0.1 < X <= 1)", 
                           "Large (0.01 < X <= 0.1)", 
                            "Medium (0.001 < X <= 0.01)", 
                            "Small (1e-04 < X <= 0.001)", 
                            "Rare (0 < X <= 1e-04)", NA))

DimPlot(list, group.by = "cloneType") + scale_color_manual(values = rev(viridis_pal(option = "B")(length(unique(list$cloneType)))))

saveRDS(list, file = "./data/processedData/PB_Fullintegrated_seuratObjects.rds")

list$filter  <- "Yes"
list$filter <- ifelse(!is.na(list$CD8.annot) | !is.na(list$CD4.annot), "No", list$filter)
list$filter <- ifelse(is.na(list$CTaa), "Yes", list$filter)

list <- subset(list, filter != "Yes")

list@meta.data$CD8.pos <- ifelse(list@assays$RNA@data["CD8A",] >=  0.4, "Yes", "No")


list <- list %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(verbose = FALSE) %>% 
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident"), 
               max.iter.harmony = 15) %>%
    FindNeighbors(reduction = "harmony", dims = 1:15) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:15,
            n.epochs = 500) %>%
    FindClusters(resolution = 0.8, 
                 algorithm = 3) %>% 
    identity()

saveRDS(list, file = "./data/processedData/PB_Fullintegrated_seuratObjects_Tcells.rds")

list <- annotateDB(list, chains = "TRB", edit.distance = 0)
list<- annotateDB(list, chains = "TRA", edit.distance = 0)

#Spike 167-180 epitope
sequence <- read.csv("./sequenceInfo/Sequences.csv")
CTaa <- as.data.frame(stringr::str_split(list$CTaa, "_", simplify = TRUE))
TA <-which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

list$TCRB.S167.180<- NA
list$TCRB.S167.180[TB] <- "Specific"

list$TCRA.S167.180 <- NA
list$TCRA.S167.180[TA] <- "Specific"


#Spike 815-830 epitope

sequence <- read.csv("./sequenceInfo/dykema.sequences.csv")
CTaa <- as.data.frame(stringr::str_split(list$CTaa, "_", simplify = TRUE))
TA <-which(CTaa$V1 %in% sequence$cdr3a)
TB <-which(CTaa$V2 %in% sequence$cdr3b)

list$TCRB.S815.830 <- NA
list$TCRB.S815.830[TB] <- "Specific"

list$TCRA.S815.830 <- NA
list$TCRA.S815.830[TA] <- "Specific"

# VDJdb sequence

sequences <- read.delim("./sequenceInfo/VDJdb.tsv")
sequences <- sequences[sequences$MHC.class == "MHCII" & sequences$Epitope.gene == "Spike",]

epitope <- c("NCTFEYVSQPFLMDL" = "S164.179",
"RISNCVADYSVLYNS" = "S356.371",
"NFSQILPDPSKPSKR" = "S800.815",
"HWFVTQRNFYEPQII" = "S1101.1115",
"VGGNYNYLYRLFRKS" = "S445.459",
"NNSYECDIPIGAGIC" = "S657.671")

list$TCRB.VDJ.db <- NA
list$TCRA.VDJ.db <- NA
for (i in seq_along(epitope)) {
  sequences$Epitope[which(sequences$Epitope == names(epitope)[i])] <- epitope[i]
  TA <- sequences[sequences$Gene == "TRA" & sequences$Epitope == epitope[i], ]
  TB <- sequences[sequences$Gene == "TRB" & sequences$Epitope == epitope[i], ]
  TA.loc <-which(CTaa$V1 %in% TA$CDR3)
  TB.loc <-which(CTaa$V2 %in% TB$CDR3)
  list$TCRB.VDJ.db[TB.loc] <- epitope[i]
  list$TCRA.VDJ.db[TA.loc] <- epitope[i]
}


list$TCRB.epitope  <- NA
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.S167.180), "S167.180", NA)
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.S815.830), "S815.830", list$TCRB.epitope)
list$TCRB.epitope <- ifelse(!is.na(list$TCRB.VDJ.db) & is.na(list$TCRB.epitope), list$TCRB.VDJ.db, list$TCRB.epitope)

list$TCRB.epitope <- ifelse(grepl("Spike", list$TRB_Epitope.target) & is.na(list$TCRB.epitope), "spike", list$TCRB.epitope)

list$TCRA.epitope  <- NA
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.S167.180), "S167.180", NA)
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.S815.830), "S815.830", list$TCRA.epitope)
list$TCRA.epitope <- ifelse(!is.na(list$TCRA.VDJ.db) & is.na(list$TCRA.epitope), list$TCRA.VDJ.db, list$TCRA.epitope)

list$TCRA.epitope <- ifelse(grepl("Spike", list$TRA_Epitope.target) & is.na(list$TCRA.epitope), "spike", list$TCRA.epitope)
dir.create("./output/Figure7")
saveRDS(list, file = "./data/processedData/PB_Fullintegrated_seuratObjects_Tcells.rds")
```

## Figure 7B

```{r}
list <- readRDS("./data/processedData/PB_Fullintegrated_seuratObjects_Tcells.rds")
spike <- read.csv("./output/Figure4/Figure4_spike.clones.csv")
cells <- unique(c(rownames(list[[]])[which(list$CTaa %in% spike$x)], 
                rownames(list[[]])[!is.na(list$TCRB.epitope)], 
                rownames(list[[]])[!is.na(list$TCRA.epitope)]))


cells <- intersect(cells, rownames(list[[]])[which(list$CD8.pos == "No")])
subset.Obj <- subset(list, cells = cells)


subset.Obj <- subset.Obj %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    quietTCRgenes() %>%
    ScaleData(verbose = FALSE) %>% 
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunHarmony(c("orig.ident")) %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    RunUMAP(reduction = "harmony",
            dims = 1:25,
            n.epochs = 500) %>%
    FindClusters(resolution = 0.5, 
                 algorithm = 3) %>% 
    identity()

subset.Obj$relative.time <- "Early"
subset.Obj$relative.time[subset.Obj$timepoint %in% c("d201", "m6", "d110", "d84")] <- "Late"

mycolors <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(subset.Obj$seurat_clusters)))

update_geom_defaults("point", list(stroke = 1))
DimPlot(subset.Obj, reduction = "umap", label = TRUE, label.size = 4, raster=FALSE, pt.size = 0.25) +
  scale_color_manual(values = mycolors) + 
  guides(colors = "none", fill = "none") + 
  theme_void() + 
  NoLegend() 

ggsave("./output/Figure7/Figure7B.png", height = 3.5, width = 4, dpi = 600, bg = "transparent")
```

## Figure 7C

```{r}
subset.Obj$study <- "s350"
subset.Obj$study[grepl("368", subset.Obj$donor)] <- "s368"
df <- data.frame(subset.Obj[[]], subset.Obj@reductions$umap@cell.embeddings)
df <- df[df$relative.time == "Late",]
ggplot(df, aes(x=UMAP_1, y = UMAP_2)) + 
  geom_point(aes(color = seurat_clusters)) + 
  scale_color_manual(values = mycolors) + 
  facet_grid(study~.) + 
  theme_void() + 
  guides(color = "none") + 
  theme(plot.title = element_blank(),
        strip.text = element_blank())
ggsave("./output/Figure7/Figure7C_1.png", height = 7, width = 4, dpi = 600)


ggplot(df, aes(x = study, fill = seurat_clusters)) + 
  geom_bar(position = "fill") + 
  scale_fill_manual(values = mycolors) + 
  theme_classic() + 
  guides(fill = "none") + 
  theme(axis.title = element_blank())
ggsave("./output/Figure7/Figure7C_2.pdf", height = 3.5, width = 4)
```

## Figure 7D

```{r}
markers <- FindAllMarkers(subset.Obj)
write.csv(markers, "./output/Figure7/Figure7_AgspecificBlood_markers.csv")

top <- markers %>%
  group_by(cluster) %>%
  slice_max(order_by = avg_log2FC, n = 8)

DotPlot(subset.Obj, features = unique(top$gene)) + 
  scale_color_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) + 
  scale_y_discrete(limits=rev)+ 
  coord_flip()
ggsave("./output/Figure7/Figure7D.pdf", height = 16, width = 6)
```


## Figure 7E

```{r}
subset.Obj <- clusterTCR(subset.Obj, 
                         group.by = "donor",
                         chain = "TRA", 
                         sequence = "aa")

mycolors <- colorRampPalette(brewer.pal(11, "RdYlBu"))(length(na.omit(unique(subset.Obj$TRA_cluster))))
names(mycolors) <- unique(subset.Obj$TRA_cluster)
DimPlot(subset.Obj, group.by = "TRA_cluster") + 
  scale_color_manual(values = mycolors) + 
  NoLegend()

df <- data.frame(subset.Obj@reductions$umap@cell.embeddings, subset.Obj@meta.data)
df <- df %>%
  group_by(TRA_cluster) %>%
  mutate(n = n())
df$TRA_cluster[df$n >= 2] <- NA

mycolors <- colorRampPalette(brewer.pal(11, "RdYlBu"))(length(na.omit(unique(df$TRA_cluster))))

ggplot(df, aes(x=UMAP_1, y = UMAP_2)) + 
  geom_point(color = "grey") + 
  geom_point(data = subset(df, !is.na(TRA_cluster)), aes(color = TRA_cluster)) + 
  scale_color_manual(values = mycolors, na.value = "grey") + 
  guides(color = "none") + 
  theme_void()
ggsave("./output/Figure7/Figure7E_1.png", height = 3.5, width = 4, dpi = 600)  

ggplot(df, aes(x=UMAP_1, y = UMAP_2)) + 
  geom_point(color = "grey") + 
  geom_point(data = subset(df, !is.na(TRA_cluster)), aes(color = TRA_cluster)) + 
  scale_color_manual(values = mycolors, na.value = "grey") + 
  guides(color = "none") + 
  facet_wrap(~donor, ncol = 5) + 
  theme_void()

ggsave("./output/Figure7/Figure7E_2.pdf", height = 3.5, width = 10)  

```


## Figure 7F

```{r}
mycolors <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(subset.Obj$seurat_clusters)))

DimPlot(subset.Obj, reduction = "umap", label.size = 4, raster=FALSE, pt.size = 0.25) +
  facet_grid(subset.Obj$study~ subset.Obj$relative.time) + 
   scale_color_manual(values = mycolors) +  
  guides(colors = "none", fill = "none") + 
  theme_void() + 
  guides(color = "none") + 
  theme(plot.title = element_blank(),
        strip.text = element_blank())
ggsave("./output/Figure7/Figure7F_1.pdf", height = 7, width = 8)


df <- data.frame(subset.Obj@reductions$umap@cell.embeddings, subset.Obj@meta.data)
df2 <- df %>%
  group_by(study, seurat_clusters, relative.time) %>%
  dplyr::count()
df2 <- df2 %>%
  group_by(study, relative.time) %>%
  mutate(sum = sum(n))
df2$proportion <- round(df2$n/df2$sum, 3)


UCB_lodes <- to_lodes_form(as.data.frame(UCBAdmissions),
                           axes = 1:3,
                           id = "Cohort")

ggplot(df2, aes(y=proportion,
                x=relative.time,
               stratum = seurat_clusters,
               alluvium = seurat_clusters)) + 
  geom_alluvium(aes(fill = seurat_clusters)) + 
  geom_stratum(aes(fill = seurat_clusters)) + 
  scale_fill_manual(values = mycolors) + 
  facet_grid(study~.) + 
  guides(fill = "none") + 
  theme_classic() + 
    theme(plot.title = element_blank(),
        strip.text = element_blank(), 
        axis.title = element_blank())

ggsave("./output/Figure7/Figure7F_2.pdf", height = 4.5, width = 5)
```

# Conclusions

```{r}
sessionInfo()
```

