---
title: "CoNGA Analysis"
author: "Nick Borcherding"
date: "11/10/2022"
output: html_document
---
```{r}
require(Seurat)
require(DropletUtils)
library(reticulate)
dat <- readRDS('./data/processedData/Fullintegrated_seuratObjects_TFH.rds')
dat <- SplitObject(dat, split.by = "orig.ident")
```

# Migrating Seurat data to CoNGA
```{r}
for (i in seq_along(dat)) {
    write10xCounts(x = dat[[i]]@assays$RNA@counts,
                   barcodes = stringr::str_split(colnames(dat[[i]]@assays$RNA@counts), "_", simplify = TRUE)[,2],
                   type = "HDF5",
                   path = paste0('./data/preCoNGA/gex/', names(dat)[i], ".h5"))
}
```

# Each contig file must undergo conversion for compatibility

Ended up writing and running the following bash script: 

```{r, eval = F}
for i in {11..118};
do
python /Users/Nick/conga/scripts/setup_10x_for_conga.py \
--filtered_contig_annotations_csvfile "/Users/Nick/Documents/GitHub/COVID_TCR/data/sequencingRuns/s""$i""/filtered_contig_annotations.csv" \
--output_clones_file "/Users/Nick/Documents/GitHub/COVID_TCR/data/preCoNGA/s""$i""_clones.tsv" \
--organism human \
--no_kpca
done
```

```{r}
new.contigs <- list.files("./data/preCoNGA/contigs")
new.contigs <- new.contigs[!grepl("barcode", new.contigs)]

samples <- stringr::str_split(new.contigs, "_", simplify = TRUE)[,1]
for(i in seq_along(new.contigs)) {
  tmp <- read.delim(paste0("./data/preCoNGA/contigs/", new.contigs[i]))
}
```

```{r, eval = FALSE}

python /Users/Nick/conga/scripts/merge_samples.py \
--samples COVID_samples2.txt \
--output_clones_file merged_COVID_clones.tsv \
--output_gex_data merged_COVID_gex.h5ad \
--organism human 
```

```{r, eval = FALSE}

python /Users/Nick/conga/scripts/run_conga.py \
--all \
--gex_data merged_COVID_gex.h5ad \
--gex_data_type h5ad \
--clones_file merged_COVID_clones.tsv \
--organism human \
--graph_vs_graph \
--outfile_prefix ./CoNGA.output \
--no_kpca
```



```{r}
ad <- anndata::read_h5ad("./data/preCoNGA/CoNGA.output_final.h5ad")

meta.data <- data.frame(ad$obs)
meta.data$CTaa <- paste0(meta.data$cdr3a, "_", meta.data$cdr3b)

meta.data <- unique(meta.data[,c("clusters_gex", "clusters_tcr", "CTaa")])

x <- unique(meta.data$CTaa[which(duplicated(meta.data$CTaa))])
output <- NULL
for(i in seq_along(x)) {
  meta.tmp <- meta.data[meta.data$CTaa == x[i],]
  clusters_gex <- (paste(str_sort(unique(meta.tmp$clusters_gex)), collapse = ";"))
  clusters_tcr <- (paste(str_sort(unique(meta.tmp$clusters_tcr)), collapse = ";"))
  out <- c(clusters_gex, clusters_tcr, x[i])
  output <- rbind(output,out)
}
output <- as.data.frame(output)
colnames(output) <- c("clusters_gex", "clusters_tcr", "CTaa")

meta.data <- meta.data[-c(which(meta.data$CTaa %in% x)),]
meta.data <- rbind(meta.data, output)
rownames(meta.data) <- meta.data$CTaa
meta.data <- meta.data[,c(1:2)]


conga <- readRDS("./data/processedData/conga_tfh_seuratObject.rds")

output <- rescaleByNeighbors(
                  list(conga[["TRA.KF"]]@cell.embeddings[,1:30], 
                      conga[["TRB.KF"]]@cell.embeddings[,1:30]),
                  k = 100)

conga[["KF.only"]] <- CreateDimReducObject(
                                embeddings = as.matrix(output),
                                stdev = rep(0, ncol(output)),
                                key = "correctedTrex",
                                jackstraw = NULL,
                                misc = list())
conga <- FindNeighbors(conga, 
                       reduction = "KF.only", 
                       graph.name = c("KF_nn", "KF_snn"))


conga <- FindClusters(conga, 
                      algorithm = 4, 
                      graph.name = "KF_snn", 
                      resolution = 0.2)


conga<- AddMetaData(conga, meta.data)
conga$clusters_gex[grepl(";",conga$clusters_gex)] <- "Multi"

conga$clusters_tcr <- factor(conga$clusters_tcr, levels = 0:15)
conga$clusters_gex <- factor(conga$clusters_gex, levels = c(0:13, "Multi"))

epitope <- stringr::str_sort(na.omit(unique(conga$TCRA.epitope)), numeric = TRUE)
        
ep.colors <- rev(brewer.pal(5, "RdYlBu"))[c(1,3,5)]
names(ep.colors) <- epitope

conga$TCRA.epitope <- factor(conga$TCRA.epitope, levels = epitope)


###################
#Direct Comparison
###################

tcrdist <- data.frame(ad$obsm$X_gex_2d, CTaa = paste0(ad$obs$cdr3a, "_", ad$obs$cdr3b))
"%!in%" <- Negate("%in%")



conga.sparse <- ad$obsp$connectivities
conga.nn.idx <- apply(conga.sparse[], 
                      1, 
                      function(x) which(x > 0) )

conga.nn.idx <- data.frame(lapply(conga.nn.idx, function(x) {
  x <- unlist(x)
  length(x) <- max(lengths(conga.nn.idx))
  return(x)
}))
conga.nn.idx <- as.data.frame(t(conga.nn.idx))
rownames(conga.nn.idx) <- NULL
conga.nn.idx$CTaa <- tcrdist$CTaa
dupl.barcodes <- conga.nn.idx$CTaa[which(duplicated(conga.nn.idx$CTaa))]
conga.nn.idx <- conga.nn.idx[conga.nn.idx$CTaa %!in% dupl.barcodes,]

conga$dupl.barcodes <- "No"
conga$dupl.barcodes[conga$CTaa %in% dupl.barcodes] <- "Yes"
subset.conga <- subset(conga, CTaa %in% conga.nn.idx$CTaa)
subset.conga <- FindNeighbors(subset.conga, 
                              reduction = "KF.only", 
                              k.param = 11, 
                              return.neighbor = TRUE)
Trex.nn.idx <- as.data.frame(subset.conga@neighbors$RNA.nn@nn.idx[,2:11])
Trex.nn.idx$CTaa <- subset.conga$CTaa

clones <- unique(Trex.nn.idx$CTaa)
output.JI <- NULL
output.DI <- NULL
for (i in seq_along(clones)) {
  
  Trex.tmp <- as.numeric(Trex.nn.idx[Trex.nn.idx$CTaa == clones[i],1:10])
  Trex.nn <- Trex.nn.idx$CTaa[Trex.tmp]
  
  cg.tmp <- as.numeric(conga.nn.idx[conga.nn.idx$CTaa == clones[i],1:10])
  conga.nn <- conga.nn.idx$CTaa[cg.tmp]
  numerator <- length(intersect(Trex.nn, conga.nn))
  denominator <- length(unique(c(conga.nn, Trex.nn)))
  if (numerator > 0) {
    sub <- length(which(conga.nn.idx$CTaa[cg.tmp] %!in% Trex.nn.idx$CTaa))
    if(sub > 0) {
      denominator <- denominator-sub
    }
  }
  JI <- numerator/denominator
  DI <- 2*(numerator)/(length(which(conga.nn.idx$CTaa[cg.tmp] %in% Trex.nn.idx$CTaa)) + length(Trex.nn))
  output.JI <- c(output.JI, JI)
  output.DI <- c(output.DI, DI)
}

JI <- data.frame(row.names = Trex.nn.idx$CTaa, JI.idx = output.JI, DI.idx = output.DI)
conga <- AddMetaData(conga, JI)


meta <- conga[[]]
tcrdist <- merge(tcrdist, meta, by = "CTaa")


#################
#Gex Based Cluster
#################
centers = tcrdist %>% 
          group_by(clusters_gex) %>% 
          summarize(across(c("X1", "X2"),mean)) %>%
          as.data.frame()

centers <-  centers[-15,]
        

gex.clusters <- ggplot(tcrdist, aes(x= X1, y = X2)) + 
                  geom_point(aes(color = clusters_gex)) +
                  geom_text(data = centers, mapping =  aes(x=centers[,2], y = centers[,3], label = clusters_gex), 
                              color = "black", size = 5) + 
                  scale_color_tableau("Tableau 20")  +
                  guides(color = "none") + 
                  xlab("gexUMAP_1") + 
                  ylab("gexUMAP_2") + 
                  theme_classic()

#################################
#Adding TCR candidate Information
#################################

confirming.seq <- list(TCRA = c("CATGLTGGGNKLTF", "CAGQLNRAAGNKLTF", "CIVPRGTGFQKLVF", "CAVRPNAGNNRKLIW", "CAVTGTYKYIF"), 
                       TCRB = c("CASSMGGGNQPQHF", "CASSPRGTEAFF", "CASSLVGGGPAEAFF", "CASSLGENDEQFF", "CASSQGNRANTEAFF")) 

tcrdist$confirmation.seq <- NA
for (i in seq_len(5)) {
  x.pos <- intersect(grep(confirming.seq[[1]][i], tcrdist$CTaa), grep(confirming.seq[[2]][i], tcrdist$CTaa))
  tcrdist$confirmation.seq[x.pos] <- paste0("TCR.", i)
}

#TCR1 = c(9,2), TCR2 = c(5,1), TCR3 = c(6,2), TCR4 = c(8,1), TCR5 = c(8,1)


epitope <- ggplot(tcrdist, aes(x= X1, y = X2)) + 
  geom_point(color = "grey") + 
  geom_point(data = subset(tcrdist, !is.na(TCRA.epitope)), 
             mapping = aes(color = TCRA.epitope)) + 
  xlab("gexUMAP_1") + 
  ylab("gexUMAP_2") + 
  scale_color_manual(values = ep.colors) + 
  theme_classic()

epitope.JI <- ggplot(tcrdist, aes(x= 1, y = JI.idx)) + 
  geom_boxplot(outlier.alpha = 0, fill = "grey") + 
  geom_jitter(data = subset(tcrdist, !is.na(TCRA.epitope)), 
             mapping = aes(color = TCRA.epitope)) + 
  scale_color_manual(values = ep.colors) + 
  theme_classic() + 
  guides(color = "none") + 
  ylab("Jacard Index of KNN") + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank())

epitope.DI <- ggplot(tcrdist, aes(x= 1, y = DI.idx)) + 
  geom_boxplot(outlier.alpha = 0, fill = "grey") + 
  geom_jitter(data = subset(tcrdist, !is.na(TCRA.epitope)), 
             mapping = aes(color = TCRA.epitope)) + 
  scale_color_manual(values = ep.colors) + 
  theme_classic() + 
  guides(color = "none") + 
  ylab("Dice Index of KNN") + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank())

Trex.cluster <- ggplot(tcrdist, aes(x= X1, y = X2)) + 
                    geom_point(aes(color = KF_snn_res.0.3)) + 
                    scale_color_tableau("Tableau 20") +
                    xlab("gexUMAP_1") + 
                    ylab("gexUMAP_2") + 
                    guides(color=guide_legend(ncol=2))   + 
                    labs(color = "Trex Cluster") + 
                    theme_classic()

conga.cluster <- ggplot(tcrdist, aes(x= X1, y = X2)) + 
  geom_point(aes(color = clusters_tcr)) + 
  scale_color_tableau("Tableau 20")  +
  guides(color=guide_legend(ncol=2)) + 
  labs(color = "CoNGA Cluster") + 
  xlab("gexUMAP_1") + 
  ylab("gexUMAP_2") + 
  theme_classic()

df1 <- data.frame(prop.table(table(tcrdist$KF_snn_res.0.2, tcrdist$TCRA.epitope, useNA = "ifany"), margin = 1))
df1.1 <- data.frame(sum = rowSums(table(tcrdist$KF_snn_res.0.2, tcrdist$TCRA.epitope, useNA = "ifany")))
df1.1$Var1 <- rownames(df1.1)
df1.1$epitope.sum <- rowSums(table(tcrdist$KF_snn_res.0.2, tcrdist$TCRA.epitope, useNA = "ifany")[,1:3])

trex.break <- ggplot(df1, aes(x= Var1, y = Freq)) + 
  geom_bar(stat = "identity", position = "fill", aes(fill = Var2)) + 
  geom_text(df1.1, mapping = aes(label = paste0(epitope.sum, " | ", sum), x = Var1, y = 1.1),  size = 2) + 
  scale_fill_manual(values = ep.colors, na.value = "grey") + 
  theme_classic() + 
  scale_x_discrete(limits = rev) + 
  ylab("Relative Proportion")  + 
  guides(fill = "none") +
  coord_flip() + 
  theme(axis.title.y = element_blank())

df2 <- data.frame(prop.table(table(tcrdist$clusters_tcr, tcrdist$TCRA.epitope, useNA = "ifany"), margin = 1))
df2.1 <- data.frame(sum = rowSums(table(tcrdist$clusters_tcr, tcrdist$TCRA.epitope, useNA = "ifany")))
df2.1$Var1 <- rownames(df2.1)
df2.1$epitope.sum <- rowSums(table(tcrdist$clusters_tcr, tcrdist$TCRA.epitope, useNA = "ifany")[,1:3])

conga.break <- ggplot(df2, aes(x= Var1, y = Freq)) + 
                geom_bar(stat = "identity", position = "fill", aes(fill = Var2)) + 
                geom_text(df2.1, mapping = aes(label = paste0(epitope.sum, " | ", sum), x = Var1, y = 1.1),  size = 2) + 
                scale_fill_manual(values = ep.colors, na.value = "grey") + 
                theme_classic() + 
                scale_x_discrete(limits = rev) + 
                labs(fill = "Spike Epitopes") + 
                ylab("Relative Proportion") + 
                guides(fill = "none") +
                coord_flip() + 
                theme(axis.title.y = element_blank())

#gex.clusters + conga.cluster +  epitope +  Trex.cluster + trex.break +  plot_layout(ncol = 3)

epitope + (epitope.JI + epitope.DI) / (trex.break + conga.break)

#ggsave("conga.epitope1.png", height = 6, width = 14)
```

# Neighborhood calculation

```{r}
library(stringr)
library(clustringr)
source("./R/sparse.merge.R")

########################################################
#Collecting TR sequences and organizing by VJ gene usage
#########################################################
TRA <- unique(data.frame(CT = stringr::str_split(conga$CTaa, "_", simplify = TRUE)[,1], 
                  VJgene = paste0( stringr::str_split(conga$CTgene, "[.]", simplify = TRUE)[,1], ".", stringr::str_split(conga$CTgene, "[.]", simplify = TRUE)[,2])))
TRA.split <- split(TRA, TRA$VJgene)


TRB <- unique(data.frame(CT = stringr::str_split(conga$CTaa, "_", simplify = TRUE)[,2], 
                  VJgene = paste0( stringr::str_split(conga$CTgene, "[.]", simplify = TRUE)[,3], ".", stringr::str_split(conga$CTgene, "[.]", simplify = TRUE)[,5])))
TRB$VJgene <- str_remove_all(TRB$VJgene, "TRAC_")
TRB$VJgene <- str_remove_all(TRB$VJgene, "_")
TRB.split <- split(TRB, TRB$VJgene)


###############################################
#Calculating edit distance across each VJ group
###############################################
#Retaining only info for chains that are <= 2 ditance apart
TRA.clust.list <- list()
for(i in seq_along(TRB.split)) {
  TRA.clust.list[[i]] <- cluster_strings(unique(TRA.split[[i]][,1]),
                                         clean = TRUE,
                                         method = "lv", 
                                         max_dist = 2)
}

TRB.clust.list <- list()

for(i in seq_along(TRB.split)) {
  TRB.clust.list[[i]] <- cluster_strings(unique(TRB.split[[i]][,1]),
                                         clean = TRUE,
                                         method = "lv", 
                                         max_dist = 2)
}

#############################################
#Merging all the groups and id the components
#############################################

TRA.adj.matrix <- list()
for(i in seq_along(TRA.clust.list)) {
  TRA.adj.matrix[[i]] <- as(as(as(TRA.clust.list[[i]]$adj_mtx, "dMatrix"), "generalMatrix"), "RsparseMatrix")
}
TRA.sparse <- merge.sparse(TRA.adj.matrix)
x <- igraph::graph_from_adjacency_matrix(TRA.sparse)
x <- simplify(x)
components <- components(x, mode = c("weak"))
out <- data.frame("cluster" = components$membership, 
                  "filtered" = names(components$membership))
write.csv(out, "./output/TRA.edit.distance.cluster.csv")
      
TRB.adj.matrix <- list()
for(i in seq_along(TRB.clust.list)) {
  TRB.adj.matrix[[i]] <- as(as(as(TRB.clust.list[[i]]$adj_mtx, "dMatrix"), "generalMatrix"), "RsparseMatrix")
}
TRB.sparse <- merge.sparse(TRB.adj.matrix)
x <- igraph::graph_from_adjacency_matrix(TRB.sparse)
x <- simplify(x)
components <- components(x, mode = c("weak"))
out <- data.frame("cluster" = components$membership, 
                  "filtered" = names(components$membership))
write.csv(out, "./output/TRB.edit.distance.cluster.csv")

####################################################
#Nearest Neighbors from the Trex-based vectorization
####################################################
conga <- FindNeighbors(conga, 
                       reduction = "KF.only", 
                       return.neighbor = TRUE)

Trex.based.nn <- conga@graphs$KF_nn

a <- igraph::graph_from_adjacency_matrix(Trex.based.nn)
plot.layout <- layout_with_mds(a)
plot(a, vertex.label = NA, layout = plot.layout)

b <- as_edgelist(a)

TRA.edit.neighbors <- read.csv("./output/TRA.edit.distance.cluster.csv")
TRB.edit.neighbors <- read.csv("./output/TRB.edit.distance.cluster.csv")

b.TRB <- b %>%
  as.data.frame()

b.TRB$V1 <- stringr::str_split(b.TRB$V1, "_", simplify = TRUE)[,2]
b.TRB$V2 <- stringr::str_split(b.TRB$V2, "_", simplify = TRUE)[,2]

unique.TRB <- unique(TRB.edit.neighbors$filtered)

edges.list.rows.to.remove <- NULL
for(i in seq_along(unique.TRB)) {
  trex.edge <- b.TRB[b.TRB[,1] == toupper(unique.TRB[i]),]
  
  edit.neighbors <- TRB.edit.neighbors[TRB.edit.neighbors$cluster == TRB.edit.neighbors[TRB.edit.neighbors$filtered == unique.TRB[i],]$cluster,]
  
  edges.list.rows.to.remove <- c(edges.list.rows.to.remove, as.numeric(rownames(trex.edge[trex.edge$V2 %!in% toupper(edit.neighbors$filtered),])))
}
edges.list.rows.to.remove <- unique(edges.list.rows.to.remove)

c <- delete.edges(a, edges.list.rows.to.remove)

plot(a, vertex.label = NA, edge.wdith = 1)
plot(c, vertex.label = NA, edge.wdith = 1)
#Need to Figure out visualizations - add patient info, clone size
```

